---
title: "Working with TCGAbiolinks package"
author: " Antonio Colaprico, Catharina Olsen, Luciano Garofano, Tiago Chedraoui Silva, Davide Garolini, Claudia Cava, Isabella Castiglioni, Houtan Noushmehr, Gianluca Bontempi, Michele Ceccarelli"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

This document describes how to search/download analyze tcga data using TCGAbiolinks package.

## TCGA downstream Analysis
 
Some examples of integration of TCGA data are shown below
```{r, eval = FALSE}

library(TCGAbiolinks)
query <- TCGAQuery(tumor = "brca",level = 3)

querySamples <- samplesfilter(query)

query <- TCGAQuery(tumor = "brca",level = 3, platform = "IlluminaHiSeq_RNASeqV2")

```

Obs: As these terms needs to be exact we validate user input before searching.

## TCGA download
You can easily download roadmap data using the `tcga.download` function.
 
```{r, eval = FALSE}
# get all samples from the query and save into tcga folder
# samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results
# samples to normalize later
TCGADownload(query,path = "data",type ="rsem.genes.results")
```

Obs: The function will structure the folders to save the data as:
_Path given by the user/Experiment folder_


## TCGA Downstream Analysis 
You can easily prepare gene expression from TCGA data using the `tcga.prepare` function.
 
```{r, eval = FALSE}
library(TCGAbiolinks)

res <- querySamples

toDown <- Reduce(intersect, list(res$Illuminahiseq_rnaseqv2, res$Genome_Wide_SNP_6, res$IlluminaGA_DNASeq))
Tumor <- 'BRCA'
downloadFolder <- "~/Downloads/Data/Samples/"
sdrfFolder <- "~/Downloads/Data/TCGAsdrf/"

#BRCArnaseqV2 <- TCGADownload(Tumor, PlatformAndAssociatedData, sdrfFolder, downloadFolder, PlatformType = 'illuminahiseq_rnaseqv2', listSample = toDown, newsample = T)

BRCAsnp<- TCGADownload(Tumor, PlatformAndAssociatedData, sdrfFolder, downloadFolder, PlatformType = 'genome_wide_snp_6', listSample = toDown, newsample = T)

#Data_CANCER_normUQ <- RnaSeqNormalization(Data_CANCER,geneInfo)
#print(paste("STEP2..Normalization for ", CancerName, " with n. ", nrow(Data_CANCER_normUQ), " genes and ", ncol(Data_CANCER_normUQ),"samples...OK" ))

#Data_CANCER_normUQ_filt <- RnaSeqFilt(Data_CANCER_normUQ,0.25)
#print(paste("STEP3..Filter quantile for ", CancerName, " with n. ", nrow(Data_CANCER_normUQ_filt), " genes and ", ncol(Data_CANCER_normUQ_filt),"samples...OK" ))
BRCArnaseqV2 <- Cancer_rnaseqv2[, toDown]

library(genefilter)
library(clue)
library(consensus)
BRCArnaseqV2MostVar <- varFilter(BRCArnaseqV2, var.func = IQR,  var.cutoff = 0.75, filterByQuantile = TRUE)

wData <- t(BRCArnaseqV2MostVar)
ddist <- dist(wData, method = "euclidean")
sHc <- hclust(ddist, method = "ward.D")
aCalinsky <- calinsky(sHc, gMax = 20)
plot(aCalinsky, type = "l", col = "grey", main = "Calinsky & Harabasz curve")
text(1:length(aCalinsky), aCalinsky, paste(1:length(aCalinsky)))

plot(sHc, labels = F, main ="Cancer cluster dendrogram all samples", xlab = "Samples with relative group color",sub="")

rect.hclust(sHc, k=3, border="red")
tabCluster <- as.matrix(cutree(sHc, k = 3))
colnames(tabCluster)<-"Cluster"
tabCluster<-cbind(Sample = rownames(tabCluster),Color = rownames(tabCluster), tabCluster)
tabCluster<-as.data.frame(tabCluster)
tabCluster<-tabCluster[order(tabCluster$Cluster,decreasing=F),]
tabCluster<-as.data.frame(tabCluster)
tabCluster$Color<-as.character(tabCluster$Color)


ccol <- palette()[1 + 1:3]

for( cc in 1:3){
  tabCluster[tabCluster[, "Cluster"] == cc, "Color"] <- ccol[cc]
}

tabCluster <- tabCluster[sHc$labels, ]

rug(which(tabCluster[sHc$order, "Color"] == "blue"), col = "blue", lwd = 3)
rug(which(tabCluster[sHc$order, "Color"] == "green3"), col = "green3", lwd = 3)
rug(which(tabCluster[sHc$order, "Color"] == "red"), col = "red", lwd = 3)

##### Differential Analisys
GroupBlueData <- BRCArnaseqV2[, as.character(tabCluster[tabCluster$Color == "blue", "Sample"])]
GroupGreen3Data <- BRCArnaseqV2[, as.character(tabCluster[tabCluster$Color == "green3", "Sample"])]
GroupRedData <- BRCArnaseqV2[, as.character(tabCluster[tabCluster$Color == "red", "Sample"])]

DEGsBlue <- DEA_edge5(cbind(GroupGreen3Data, GroupRedData), GroupBlueData, "GroupOther", "GroupBlue")
DEGsGreen3 <- DEA_edge5(cbind(GroupBlueData, GroupRedData), GroupGreen3Data, "GroupOther", "GroupGreen3")
DEGsRed <- DEA_edge5(cbind(GroupBlueData, GroupGreen3Data), GroupRedData, "GroupOther", "GroupRed")

DEGsBlueLevel <- CreateTabLevel( as.matrix(rownames(DEGsBlue)),DEGsBlue,"GroupBlue","GroupOther",GroupBlueData,cbind(GroupGreen3Data, GroupRedData),typeOrder=T)
DEGsGreen3Level <- CreateTabLevel( as.matrix(rownames(DEGsGreen3)),DEGsGreen3,"GroupGreen3","GroupOther",GroupGreen3Data,cbind(GroupBlueData, GroupRedData),typeOrder=T)
DEGsRedLevel <- CreateTabLevel( as.matrix(rownames(DEGsRed)),DEGsRed,"GroupRed","GroupOther",GroupRedData,cbind(GroupBlueData, GroupGreen3Data),typeOrder=T)

blueDEGs <- DEGsBlueLevel[DEGsBlueLevel$FDR < 0.01 & DEGsBlueLevel$logFC >= 1, ]
blueDEGs <- blueDEGs[order(blueDEGs$FDR), ]
green3DEGs <- DEGsGreen3Level[DEGsGreen3Level$FDR < 0.01 & DEGsGreen3Level$logFC >= 1, ]
green3DEGs <- green3DEGs[order(green3DEGs$FDR), ]
redDEGs <- DEGsRedLevel[DEGsRedLevel$FDR < 0.01 & DEGsRedLevel$logFC >= 1, ]
redDEGs <- redDEGs[order(redDEGs$FDR), ]

blueDEGsSpec <- blueDEGs[setdiff(rownames(blueDEGs), union(rownames(green3DEGs), rownames(redDEGs))), ]
green3DEGsSpec <- green3DEGs[setdiff(rownames(green3DEGs), union(rownames(blueDEGs), rownames(redDEGs))), ]
redDEGsSpec <- redDEGs[setdiff(rownames(redDEGs), union(rownames(blueDEGs), rownames(green3DEGs))), ]

blueDEGsSpec <- blueDEGsSpec[1:50, ]
green3DEGsSpec <- green3DEGsSpec[1:50, ]
redDEGsSpec <- redDEGsSpec[1:50, ]

tabCluster <- tabCluster[order(tabCluster$Color), ]

MfiltQuantileOrdered <- BRCArnaseqV2[c(rownames(blueDEGsSpec), rownames(green3DEGsSpec), rownames(redDEGsSpec)), rownames(tabCluster)]
#MfiltQuantileOrdered <- (MfiltQuantileOrdered - rowMeans(MfiltQuantileOrdered))/apply(MfiltQuantileOrdered, 1, sd)


MRactivity <- t(MfiltQuantileOrdered)
HMactivity <- t(quantileNormalization(t(MRactivity)))

summary(as.vector(HMactivity))
quantile(HMactivity, 0.15); quantile(HMactivity, 0.85)
HMactivity[HMactivity <= quantile(HMactivity, 0.15)] <- quantile(HMactivity, 0.15)
HMactivity[HMactivity >= quantile(HMactivity, 0.85)] <- quantile(HMactivity, 0.85)

column_annotation <- matrix(" ", nrow = nrow(HMactivity), ncol = 1)
column_annotation[, 1] <- tabCluster$Color

row_annotation <- matrix(" ", nrow = 1, ncol = ncol(HMactivity))
row_annotation[1, ] <- c(rep("blue", nrow(blueDEGsSpec)), rep("green3", nrow(green3DEGsSpec)), rep("red", nrow(redDEGsSpec)))



#png("tmp.png", width = 4096, height = 3200)
heatmap.3(t(HMactivity), ColSideColors = column_annotation, RowSideColors = row_annotation,
          key = FALSE, Colv = NA, Rowv = NA,
          scale = "none", col = greenred(75), dendrogram = "none",
          labRow = NA, labCol = NA,
          margins = c(1, 6), side.height.fraction = 0.25, keysize = 0.4, cexRow = .6)
#dev.off()


BRCAcnv <- TCGADownload(Tumor, PlatformAndAssociatedData, sdrfFolder, downloadFolder, PlatformType = 'genome_wide_snp_6', listSample = toDown, newsample = F)

```

## Methylation analysis

Some downstream analysis from methylation data can be done with TCGAbiolinks.
An example is shown below. We search, download and prepare data from the 
HumanMethylation450 platform.

```{r, eval = FALSE}
library(TCGAbiolinks)
samples <- c("TCGA-06-0125-01A-01D-A45W-05","TCGA-06-0152-01A-02D-A45W-05",
             "TCGA-06-0171-01A-02D-A45W-05","TCGA-06-0190-01A-01D-A45W-05",
             "TCGA-06-0210-01A-01D-A45W-05","TCGA-06-0211-01A-01D-A45W-05",
             "TCGA-06-0221-01A-01D-A45W-05","TCGA-14-0736-01A-01D-A45W-05",
             "TCGA-19-0957-01C-01D-A45W-05","TCGA-19-1389-01A-01D-A45W-05",
             "TCGA-14-1402-01A-01D-A45W-05","TCGA-19-4065-01A-01D-2004-05",
             "TCGA-06-0125-02A-11D-2004-05","TCGA-06-0152-02A-01D-2004-05",
             "TCGA-06-0171-02A-11D-2004-05","TCGA-06-0190-02A-01D-2004-05",
             "TCGA-06-0210-02A-01D-2004-05","TCGA-06-0211-02A-02D-2004-05",
             "TCGA-06-0221-02A-11D-2004-05","TCGA-14-0736-02A-01D-2004-05",
             "TCGA-19-0957-02A-11D-2004-05","TCGA-19-1389-02A-21D-2004-05",
             "TCGA-14-1402-02A-01D-2004-05","TCGA-19-4065-02A-11D-2004-05")

query <- TCGAQuery(tumor = "gbm", platform = "HumanMethylation450",
                   level = "3", samples = samples)
TCGADownload(query,path = "exampleData")

#-------------------------- Preparing data and metadata
# met: data frame with probe info (probe id, ge) and beta values
# beta: data frame with only beta values (probe vs patient)
# probe: data frame with only probe info(probe vs patient)
# Conclusion: met = probe +d beta
# met.md = patient metadata
met <- TCGAPrepare(query, dir="exampleData")
probe  <- met[,1:4]
beta   <- met[,5:ncol(met)]


query <- TCGAQuery(tumor = "gbm", platform = "bio", level = "2")
TCGADownload(query, path = "exampleData")
met.md <- TCGAPrepare(query, dir="exampleData")

samples <- colnames(beta)
idx <- is.element(met.md$bcr_patient_barcode,strtrim(samples,12))
met.md <- met.md[idx,]
```

After we divided the data in groups 
in order to analyse the data. Using the metadata, it is possible to create an
survival plot with the function survivalPlot as follows:

```{r, eval = FALSE}
# random split of pacients into groups
met.md$cluster <- c(rep("group1",nrow(met.md)/4),
                    rep("group2",nrow(met.md)/4),
                    rep("group3",nrow(met.md)/4),
                    rep("group4",nrow(met.md)-3*(floor(nrow(met.md)/4))))
rownames(met.md) <- met.md$bcr_patient_barcode

#---------------------- survival
survivalPlot(met.md)
```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("survival.png")
grid.raster(img)
```

Using the data and calculating the mean methylation per groups it is possible to create an
mean methylation boxplot with the function met.mean.boxplot as follows:

```{r, eval = FALSE}
#----------------------mean methylation
met.mean <- data.frame(apply(beta,2,mean,na.rm=TRUE))
colnames(met.mean) <- "avg"
met.mean$patient <-  strtrim(rownames(met.mean),12)
aux <- merge(met.mean,met.md, by.x="patient",by.y="bcr_patient_barcode")
met.mean.boxplot(aux)
```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("meanmet.png")
grid.raster(img)
```

The final example is a volcano plot 
that compares the methylation data between the two groups.

```{r, eval = FALSE}
#----------------------- calculate.pvalues (just an example)
beta.t <-  data.frame(t(beta))
# TODO verify class of the object
pvalues <- calculate.pvalues(beta.t,c(1,3,5,7,9,11),c(2,4,6,8,10,12))

#------------- volcano plot
probe$p.value <- pvalues[,1]
probe$p.value.adj <- pvalues[,2]
diffmean <- diffmean(beta[,c(1,3,5,7,9,11)],beta[,c(2,4,6,8,10,12)])
probe <- cbind(probe,diffmean)
hypo.hyper <- volcanoPlot(data, p.cut = 0.85)
```

The output will be an graph as the figure below. The function will 
return the values that are more relevant.

```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("volcano.png")
grid.raster(img)
```

The next part, we will analyse rxpression data and methylation data with 
starburst analysis.
```{r, eval = FALSE}
 FDR <- runif(nrow(met),0,1)
 p.value.adj <- runif(nrow(met),0,1)
 GeneSymbol <- met$Gene_Symbol
 dm <-  runif(nrow(met),-3,3)
 expression <- data.frame(GeneSymbol,FDR,p.value.adj,dm)
 gene.met <- starbursAnalysis(met,expression)
 starburstPlot(gene.met)
```
