---
title: "Working with TCGAbiolinks package"
author: " Antonio Colaprico, Tiago Chedraoui Silva, Luciano Garofano, Catharina Olsen, Davide Garolini, Claudia Cava, Isabella Castiglioni, Houtan Noushmehr, Gianluca Bontempi, Michele Ceccarelli"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

This document describes how to search, download and analyze TCGA data using the TCGAbiolinks package.

## TCGA query
 You can easily search TCGA samples using the `tcgaQuery` function.
 Using a summary of filters as used in the TCGA portal, the function works with the following 
 parameters:

* tumor
* platform
* added.since
* added.up.to
* listSample
* level
* center

Some search examples are shown below
```{r, eval = FALSE}

library(TCGAbiolinks)
query <- TCGAQuery(tumor = "brca",level = 3)
querySamples <- samplesfilter(query)
query <- TCGAQuery(tumor = "brca",level = 3, platform = "IlluminaHiSeq_RNASeqV2")

```

The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)

query <- TCGAQuery(tumor = "brca",level = 3)
querySamples <- samplesfilter(query)

matSamples <- matrix(0,length(querySamples),length(querySamples))
colnames(matSamples) <- names(querySamples)
rownames(matSamples) <- names(querySamples)

for ( i in 1: nrow(matSamples)){
    for( j in 1: nrow(matSamples)){
             matSamples[i,j] <- length(intersect(querySamples[[i]], querySamples[[j]]))
    }
    }


colnames(matSamples) <- paste("p",1:length(names(querySamples)), names(querySamples),sep=".")
rownames(matSamples) <- paste("p",1:length(names(querySamples)), names(querySamples),sep=".")
colnames(matSamples) <- substr( colnames(matSamples), 1,4)

knitr::kable(matSamples, digits = 2, caption = "Table common samples among platforms from TCGAQuery", row.names = TRUE)
```
\newpage

Some examples of TCGAQuery are shown below
```{r, eval = FALSE}
query <- TCGAQuery(tumor = "gbm",
                    added.since = "01/12/2013",
                    added.up.to = "06/01/2014")

query <- TCGAQuery(tumor = c("gbm","lgg"),
                    platform = c("HumanMethylation450","HumanMethylation27"))

query <- TCGAQuery(tumor = "gbm", platform = "bio")

query <- TCGAQuery(tumor = "gbm", platform = "HumanMethylation450", level = "3")

query <- TCGAQuery(samples = "TCGA-61-1743-01A-01D-*")

query <- TCGAQuery(samples = "TCGA-61-1743-01A-01D-0649-04")

query <- TCGAQuery(samples = "TCGA-61-1743-01A-01D-0649-04", level = 3)

query <- TCGAQuery(samples = "TCGA-61-1743-01A-01D-0649-04",
                    tumor = "OV", platform = "CGH-1x1M_G4447A")
```




Query version about a platform for example IlluminaHiSeq_RNASeqV2
```{r, eval = FALSE}
library(TCGAbiolinks)

BRCA_RNASeqV2_version <- TCGAVersion(Tumor = "brca", PlatformType = "illuminahiseq_rnaseqv2")

```

The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
BRCA_RNASeqV2_version <- BRCA_RNASeqV2_version[order(BRCA_RNASeqV2_version$SizeMbyte,decreasing=T),]
knitr::kable(BRCA_RNASeqV2_version, digits = 2, caption = "Table with version, number of samples and size (Mbyte) of BRCA IlluminaHiSeq_RNASeqV2 Level 3", row.names = FALSE)
```

## TCGA query, download, prepare list of a barcode
You can easily search TCGA samples, download and prepare a matrix of gene expression.
```{r, eval = FALSE}

# You can define a list of samples to query and download providing relative TCGA barcodes.
listSamples <- c("TCGA-33-6737-01","TCGA-33-6737-11","TCGA-43-6143-01","TCGA-43-6143-11",
                 "TCGA-43-6647-01","TCGA-43-6647-11","TCGA-43-6771-01","TCGA-43-6771-11")

# Query all available platforms with a list of barcode 
query <- TCGAQuery(samples = listSamples)

# Query platform IlluminaHiSeq_RNASeqV2 with a list of barcode 
query <- TCGAQuery(tumor = "lusc", samples = listSamples, 
  platform = "IlluminaHiSeq_RNASeqV2", level = "3")

# Download a list of barcodes with platform IlluminaHiSeq_RNASeqV2
TCGADownload(query, path = "dataLusc", type = "rsem.genes.results", samples = listSamples)

# Prepare expression matrix with gene id in rows and samples (barcode) in columns
# rsem.genes.results as values
LuscMatrix <- TCGAPrepare(query,"dataLusc")
```

\newpage

The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
dataGE <- dataBRCA[sample(rownames(dataBRCA),10),sample(colnames(dataBRCA),7)]

knitr::kable(dataGE, digits = 2, caption = "Example of a matrix of gene expression (10 genes in rows and 7 samples in columns)",row.names = TRUE)
```


## TCGA download
You can easily download data using the `tcga.download` function.
 
```{r, eval = FALSE}
# get all samples from the query and save them in the tcga folder
# samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results
# samples to normalize later
TCGADownload(query, path = "data", type = "rsem.genes.results")

```

Comment: The function will structure the folders to save the data as:
_Path given by the user/Experiment folder_

## TCGA download clinic data
 
```{r, eval = FALSE}

# Function clinic
# clinic(cancer,clinical_data_type)
# input: cancer, clinical_data_type 
# cancer = a character vector indicating cancer type to download
# Options include:
# ACC,BLCA,BRCA,CESC,COAD,DLBC,ESCA,GBM
# HNSC,KICH,KIRC,KIRP,LAML,LGG,LIHC,LUAD,
# LUSC,OV,PAAD,PRAD,READ,SARC,SKCM,STAD,THCA
# UCEC,UCS

# For information about cancer types: https://tcga-data.nci.nih.gov/tcga/
#
# clinical_data_type = a character vector indicating the types of 
# clinical data 
# 
# Options include:
# biospecimen_aliquot, biospecimen_analyte, biospecimen_cqcf, biospecimen_diagnostic_slides,
# biospecimen_normal_control, biospecimen_portion, biospecimen_protocol,
# biospecimen_sample, biospecimen_shipment_portion, biospecimen_slide,
# biospecimen_tumor_sample, clinical_cqcf, clinical_drug, clinical_follow_up_v1.5,
# clinical_follow_up_v2.1, clinical_follow_up_v4.0, clinical_follow_up_v4.0_nte,
# clinical_nte, clinical_omf_v4.0, clinical_patient, clinical_radiation 
# 
                       
clinic("gbm","clinical_patient")

```


## TCGA Downstream Analysis TCGAprepare, TCGAanalyze, TCGAvisualize
You can prepare gene expression from TCGA data using the `tcga.prepare` function.
 
```{r, eval = FALSE}
# Downstream analysis using gene expression data  
# TCGA samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results

# save(DAVID_BP_matrix, DAVID_CC_matrix, DAVID_MF_matrix, IPAGenes, listIPA_pathways,
#   file = "dataEnrichmentAnalysis.rda")
# save(dataBRCA, geneInfo , file = "dataGeneExpression.rda")
library(TCGAbiolinks)

# normalization of genes
dataNorm <- TCGAbiolinks::RnaSeqNormalization(dataBRCA, geneInfo)

# quantile filter of genes
dataFilt <- RnaSeqFilt(dataNorm, 0.25)

# selection of normal samples "NT"
samplesNT <- MultiSampleTypes(colnames(dataFilt), typesample = c("NT"))

# selection of tumor samples "TP"
samplesTP <- MultiSampleTypes(colnames(dataFilt), typesample = c("TP"))


```

## TCGA analyze and visualize (Diff.expr.analysis (DEA))
You can analyze gene expression from TCGA data using the `tcga.analyze` function.
 
```{r, eval = FALSE}
# Downstream analysis using gene expression data  
# TCGA samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results
# save(dataBRCA, geneInfo , file = "dataGeneExpression.rda")
library(TCGAbiolinks)

# Diff.expr.analysis (DEA)
dataDEGs <- DEArnaSEQ(dataFilt[,samplesNT], dataFilt[,samplesTP], "Normal", "Tumor")

# DEGs filter by abs(logFC) >=1
dataDEGsFilt <- dataDEGs[abs(dataDEGs$logFC) >= 1,]

# DEGs table with expression values in normal and tumor samples
dataDEGsFiltLevel <- CreateTabLevel(dataDEGsFilt,"Tumor","Normal",dataFilt[,samplesTP],dataFilt[,samplesNT])

```
The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
dataDEGsFiltLevel$FDR <- format(dataDEGsFiltLevel$FDR, scientific = TRUE)
knitr::kable(dataDEGsFiltLevel[1:10,], digits = 2, caption = "Table DEGs after DEA",row.names = FALSE)
```



## TCGA analyze and visualize (Enrichment Analysis)
You can analyze gene expression from TCGA data using the `tcga.analyze` function.
 
```{r, eval = FALSE}

# Enrichment Analysis EA
# Gene Ontology (GO) and Pathway enrichment by DEGs list
Genelist <- rownames(dataDEGsFiltLevel)

TimeUse(ansEA <- EAcomplete(TFname="DEA genes Normal Vs Tumor",Genelist))

# Enrichment Analysis EA (TCGAVisualize)
# Gene Ontology (GO) and Pathway enrichment IPA barPlot

IPAbarplot(tf = rownames(ansEA$ResBP), 
           GOBPTab = ansEA$ResBP,
           GOCCTab = ansEA$ResCC,
           GOMFTab = ansEA$ResMF,
           PathTab = ansEA$ResPat,
           nRGTab = Genelist, 
           nBar = 10)




```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("EAplot.png")
grid.raster(img)
```

## TCGA analyze (Enrichment Analysis)
You can analyze gene expression from TCGA data using the `tcga.analyze` function.
 
```{r, eval = FALSE}
library(TCGAbiolinks)

# normalization of genes
dataNorm <- TCGAbiolinks::RnaSeqNormalization(dataBRCA, geneInfo)

# quantile filter of genes
dataFilt <- RnaSeqFilt(dataNorm, 0.25)

# Principal Component Analysis plot for ntop selected DEGs
plotPCAforGroups(dataFilt,dataDEGsFiltLevel, ntopgenes = 200)

# boxplot of normalized data
#sampleGenes <- rownames(dataDEGsFilt[dataDEGsFilt$logFC >=1,])[1:20]
#boxplot(log(dataBRCA[sampleGenes,]), las = 2)
#boxplot(log(dataFilt[sampleGenes,]), las = 2)


```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("PCAtop200DEGs.png")
grid.raster(img)
```



## TCGA investigate
Find most studied TFs in pubmed related to a specific cancer, disease, or tissue
 
```{r, eval = FALSE}
# First perform DEGs with TCGAanalyze
# See previous section 
library(TCGAbiolinks)

# Select only transcription factors (TFs) from DEGs
TFs <- IPAGenes[IPAGenes$Family =="transcription regulator",]
TFs_inDEGs <- intersect(TFs$Gene, dataDEGsFiltLevel$mRNA )
dataDEGsFiltLevelTFs <- dataDEGsFiltLevel[TFs_inDEGs,]

# Order table DEGs TFs according to Delta decrease
dataDEGsFiltLevelTFs <- dataDEGsFiltLevelTFs[order(dataDEGsFiltLevelTFs$Delta,decreasing = TRUE),]

# Find Pubmed of TF studied related to cancer
tabDEGsTFPubmed <- TCGAinvestigate("breast", dataDEGsFiltLevelTFs, topgenes = 10)
                 
```

The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
knitr::kable(tabDEGsTFPubmed, digits = 2, caption = "Table with most studied TF in pubmed related to a specific cancer",row.names = FALSE)
```


## TCGA downstream Analysis
 
Some examples for the integration of TCGA data are shown below
```{r, eval = FALSE}

library(TCGAbiolinks)
query <- TCGAQuery(tumor = "brca",level = 3)

querySamples <- samplesfilter(query)

query <- TCGAQuery(tumor = "brca",level = 3, platform = "IlluminaHiSeq_RNASeqV2")

```

Comment: As these terms need to be exact we validate user input before searching.

## TCGA download
You can easily download roadmap data using the `tcga.download` function.
 
```{r, eval = FALSE}
# get all samples from the query and save into tcga folder
# samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results
# samples to normalize later
TCGADownload(query,path = "data",type ="rsem.genes.results")
```

Obs: The function will structure the folders to save the data as:
_Path given by the user/Experiment folder_


## TCGA Downstream Analysis Integration 
You can easily prepare gene expression from TCGA data using the `tcga.prepare` function.
 
```{r, eval = FALSE}
library(TCGAbiolinks)

#res <- querySamples

#toDown <- Reduce(intersect, list(res$Illuminahiseq_rnaseqv2, res$Genome_Wide_SNP_6, res$IlluminaGA_DNASeq))
#Tumor <- 'BRCA'
#downloadFolder <- "~/Downloads/Data/Samples/"
#sdrfFolder <- "~/Downloads/Data/TCGAsdrf/"

#BRCArnaseqV2 <- TCGADownload(Tumor, PlatformAndAssociatedData, sdrfFolder, downloadFolder,
#   PlatformType = 'illuminahiseq_rnaseqv2', listSample = toDown, newsample = TRUE)

#BRCAsnp<- TCGADownload(Tumor, PlatformAndAssociatedData, sdrfFolder, downloadFolder,
#   PlatformType = 'genome_wide_snp_6', listSample = toDown, newsample = TRUE)

#Data_CANCER_normUQ <- RnaSeqNormalization(Data_CANCER,geneInfo)
#Data_CANCER_normUQ_filt <- RnaSeqFilt(Data_CANCER_normUQ,0.25)

#BRCArnaseqV2 <- Cancer_rnaseqv2[, toDown]

library(genefilter)
library(clue)
#library(consensus)

BRCArnaseqV2 <- dataBRCA
BRCArnaseqV2MostVar <- varFilter(BRCArnaseqV2, var.func = IQR,  var.cutoff = 0.75, filterByQuantile = TRUE)

wData <- t(BRCArnaseqV2MostVar)
ddist <- dist(wData, method = "euclidean")
sHc <- hclust(ddist, method = "ward.D")
#aCalinsky <- calinsky(sHc, gMax = 20)
#plot(aCalinsky, type = "l", col = "grey", main = "Calinsky & Harabasz curve")
#text(1:length(aCalinsky), aCalinsky, paste(1:length(aCalinsky)))

plot(sHc, labels = FALSE, main ="BRCA Cancer cluster dendrogram all samples",
   xlab = "Samples with relative group color",sub="")

rect.hclust(sHc, k=3, border="red")
tabCluster <- as.matrix(cutree(sHc, k = 3))
colnames(tabCluster)<-"Cluster"
tabCluster<-cbind(Sample = rownames(tabCluster),Color = rownames(tabCluster), tabCluster)
tabCluster<-as.data.frame(tabCluster)
tabCluster<-tabCluster[order(tabCluster$Cluster,decreasing = FALSE),]
tabCluster<-as.data.frame(tabCluster)
tabCluster$Color<-as.character(tabCluster$Color)


ccol <- palette()[1 + 1:3]

for( cc in 1:3){
  tabCluster[tabCluster[, "Cluster"] == cc, "Color"] <- ccol[cc]
}

tabCluster <- tabCluster[sHc$labels, ]

rug(which(tabCluster[sHc$order, "Color"] == "blue"), col = "blue", lwd = 3)
rug(which(tabCluster[sHc$order, "Color"] == "green3"), col = "green3", lwd = 3)
rug(which(tabCluster[sHc$order, "Color"] == "red"), col = "red", lwd = 3)
```

The result is shown below:
```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("BRCA_cluster.png")
grid.raster(img)
```


```{r, eval = FALSE}
library(TCGAbiolinks)

##### Differential analysis
GroupBlueData <- BRCArnaseqV2[, as.character(tabCluster[tabCluster$Color == "blue", "Sample"])]
GroupGreen3Data <- BRCArnaseqV2[, as.character(tabCluster[tabCluster$Color == "green3", "Sample"])]
GroupRedData <- BRCArnaseqV2[, as.character(tabCluster[tabCluster$Color == "red", "Sample"])]

DEGsBlue <- DEArnaSEQ(cbind(GroupGreen3Data, GroupRedData), GroupBlueData, "GroupOther", "GroupBlue")
DEGsGreen3 <- DEArnaSEQ(cbind(GroupBlueData, GroupRedData), GroupGreen3Data, "GroupOther", "GroupGreen3")
DEGsRed <- DEArnaSEQ(cbind(GroupBlueData, GroupGreen3Data), GroupRedData, "GroupOther", "GroupRed")


dataDEGs <- DEArnaSEQ(dataFilt[,samplesNT], dataFilt[,samplesTP], "Normal", "Tumor")

# DEGs filter by abs(logFC) >=1
dataDEGsFilt <- dataDEGs[abs(dataDEGs$logFC) >= 1,]

dataDEGsFiltLevel <- CreateTabLevel(dataDEGsFilt,"Tumor","Normal",dataFilt[,samplesTP],dataFilt[,samplesNT])

DEGsBlueLevel <- CreateTabLevel(DEGsBlue, "GroupBlue", "GroupOther", 
  GroupBlueData, cbind(GroupGreen3Data, GroupRedData), typeOrder = TRUE)
DEGsGreen3Level <- CreateTabLevel(DEGsGreen3, "GroupGreen3", "GroupOther",
  GroupGreen3Data, cbind(GroupBlueData, GroupRedData), typeOrder = TRUE)
DEGsRedLevel <- CreateTabLevel(DEGsRed, "GroupRed", "GroupOther",
  GroupRedData, cbind(GroupBlueData, GroupGreen3Data), typeOrder = TRUE)

blueDEGs <- DEGsBlueLevel[DEGsBlueLevel$FDR < 0.01 & DEGsBlueLevel$logFC >= 1, ]
blueDEGs <- blueDEGs[order(blueDEGs$FDR), ]
green3DEGs <- DEGsGreen3Level[DEGsGreen3Level$FDR < 0.01 & DEGsGreen3Level$logFC >= 1, ]
green3DEGs <- green3DEGs[order(green3DEGs$FDR), ]
redDEGs <- DEGsRedLevel[DEGsRedLevel$FDR < 0.01 & DEGsRedLevel$logFC >= 1, ]
redDEGs <- redDEGs[order(redDEGs$FDR), ]

blueDEGsSpec <- blueDEGs[setdiff(rownames(blueDEGs), union(rownames(green3DEGs), rownames(redDEGs))), ]
green3DEGsSpec <- green3DEGs[setdiff(rownames(green3DEGs), union(rownames(blueDEGs), rownames(redDEGs))), ]
redDEGsSpec <- redDEGs[setdiff(rownames(redDEGs), union(rownames(blueDEGs), rownames(green3DEGs))), ]

blueDEGsSpec <- blueDEGsSpec[1:50, ]
green3DEGsSpec <- green3DEGsSpec[1:50, ]
redDEGsSpec <- redDEGsSpec[1:50, ]

tabCluster <- tabCluster[order(tabCluster$Color), ]

MfiltQuantileOrdered <- BRCArnaseqV2[c(rownames(blueDEGsSpec), rownames(green3DEGsSpec), 
  rownames(redDEGsSpec)), rownames(tabCluster)]
#MfiltQuantileOrdered <- (MfiltQuantileOrdered - rowMeans(MfiltQuantileOrdered))/apply(MfiltQuantileOrdered,
# 1, sd)


MRactivity <- t(MfiltQuantileOrdered)
#HMactivity <- t(quantileNormalization(t(MRactivity)))

HMactivity <- MRactivity
thresholdquantile<-0.75
HMactivity[ HMactivity >= quantile(HMactivity,thresholdquantile)]<- quantile(HMactivity,thresholdquantile)


summary(as.vector(HMactivity))
quantile(HMactivity, 0.15); quantile(HMactivity, 0.85)
HMactivity[HMactivity <= quantile(HMactivity, 0.15)] <- quantile(HMactivity, 0.15)
HMactivity[HMactivity >= quantile(HMactivity, 0.85)] <- quantile(HMactivity, 0.85)

column_annotation <- matrix(" ", nrow = nrow(HMactivity), ncol = 1)
column_annotation[, 1] <- tabCluster$Color

row_annotation <- matrix(" ", nrow = 1, ncol = ncol(HMactivity))
row_annotation[1, ] <- c(rep("blue", nrow(blueDEGsSpec)), rep("green3", nrow(green3DEGsSpec)), 
  rep("red", nrow(redDEGsSpec)))



library("GMD")

png("BRCA_heatmap.png", width = 1200, height = 800)
heatmap.3(t(HMactivity), ColSideColors = column_annotation, RowSideColors = row_annotation,
          key = FALSE, Colv = NA, Rowv = NA,
          scale = "none", 
          #col = greenred(75), 
          dendrogram = "none",
          #labRow = NA, labCol = NA,
          margins = c(1, 6), side.height.fraction = 0.25, keysize = 1.4, cexRow = 1.6)
dev.off()

#BRCAcnv <- TCGADownload(Tumor, PlatformAndAssociatedData, sdrfFolder, downloadFolder,
#   PlatformType = 'genome_wide_snp_6', listSample = toDown, newsample = FALSE)

```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("BRCA_heatmap.png")
grid.raster(img)
```


#### TCGA downstream methylation analysis

Some downstream analysis from methylation data can be done with TCGAbiolinks.
An example is shown below. We search, download and prepare data from the 
HumanMethylation450 platform.

```{r, eval = FALSE}
library(TCGAbiolinks)
samples <- c("TCGA-06-0125-01A-01D-A45W-05","TCGA-06-0152-01A-02D-A45W-05",
             "TCGA-06-0171-01A-02D-A45W-05","TCGA-06-0190-01A-01D-A45W-05",
             "TCGA-06-0210-01A-01D-A45W-05","TCGA-06-0211-01A-01D-A45W-05",
             "TCGA-06-0221-01A-01D-A45W-05","TCGA-14-0736-01A-01D-A45W-05",
             "TCGA-19-0957-01C-01D-A45W-05","TCGA-19-1389-01A-01D-A45W-05",
             "TCGA-14-1402-01A-01D-A45W-05","TCGA-19-4065-01A-01D-2004-05",
             "TCGA-06-0125-02A-11D-2004-05","TCGA-06-0152-02A-01D-2004-05",
             "TCGA-06-0171-02A-11D-2004-05","TCGA-06-0190-02A-01D-2004-05",
             "TCGA-06-0210-02A-01D-2004-05","TCGA-06-0211-02A-02D-2004-05",
             "TCGA-06-0221-02A-11D-2004-05","TCGA-14-0736-02A-01D-2004-05",
             "TCGA-19-0957-02A-11D-2004-05","TCGA-19-1389-02A-21D-2004-05",
             "TCGA-14-1402-02A-01D-2004-05","TCGA-19-4065-02A-11D-2004-05")

query <- TCGAQuery(tumor = "gbm", platform = "HumanMethylation450",
                   level = "3", samples = samples)
TCGADownload(query,path = "exampleData")

#-------------------------- Preparing data and metadata
# met: data frame with probe info (probe id, ge) and beta values
# beta: data frame with only beta values (probe vs patient)
# probe: data frame with only probe info(probe vs patient)
# Conclusion: met = probe +d beta
# met.md = patient metadata
met <- TCGAPrepare(query, dir="exampleData")
probe  <- met[,1:4]
beta   <- met[,5:ncol(met)]


query <- TCGAQuery(tumor = "gbm", platform = "bio", level = "2")
TCGADownload(query, path = "exampleData")
met.md <- TCGAPrepare(query, dir="exampleData")

samples <- colnames(beta)
idx <- is.element(met.md$bcr_patient_barcode,strtrim(samples,12))
met.md <- met.md[idx,]
```

We divided the data into groups 
in order to analyze the data. Using the metadata, it is possible to create a
survival plot with the function survivalPlot as follows:

```{r, eval = FALSE}
# random split of pacients into groups
met.md$cluster <- c(rep("group1",nrow(met.md)/4),
                    rep("group2",nrow(met.md)/4),
                    rep("group3",nrow(met.md)/4),
                    rep("group4",nrow(met.md)-3*(floor(nrow(met.md)/4))))
rownames(met.md) <- met.md$bcr_patient_barcode

#---------------------- survival
survivalPlot(met.md)
```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("survival.png")
grid.raster(img)
```

Using the data and calculating the mean methylation per group, it is possible to create a
mean methylation boxplot with the function met.mean.boxplot as follows:

```{r, eval = FALSE}
#----------------------mean methylation
met.mean <- data.frame(apply(beta, 2, mean, na.rm = TRUE))
colnames(met.mean) <- "avg"
met.mean$patient <- strtrim(rownames(met.mean), 12)
aux <- merge(met.mean, met.md, by.x = "patient", by.y = "bcr_patient_barcode")
met.mean.boxplot(aux)
```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("meanmet.png")
grid.raster(img)
```

The final example is a volcano plot 
that compares the methylation data between the two groups.

```{r, eval = FALSE}
#----------------------- calculate.pvalues (just an example)
beta.t <-  data.frame(t(beta))
# TODO verify class of the object
pvalues <- calculate.pvalues(beta.t,c(1,3,5,7,9,11),c(2,4,6,8,10,12))

#------------- volcano plot
probe$p.value <- pvalues[,1]
probe$p.value.adj <- pvalues[,2]
diffmean <- diffmean(beta[,c(1,3,5,7,9,11)],beta[,c(2,4,6,8,10,12)])
probe <- cbind(probe,diffmean)
hypo.hyper <- volcanoPlot(data, p.cut = 0.85)
```

The output will be an graph such as the figure below. The function will 
return the most relevant values.

```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("volcano.png")
grid.raster(img)
```

In the next part, we will analyze expression data and methylation data with 
starburst analysis.
```{r, eval = FALSE}
 FDR <- runif(nrow(met), 0, 1)
 p.value.adj <- runif(nrow(met), 0, 1)
 GeneSymbol <- met$Gene_Symbol
 dm <- runif(nrow(met), -3, 3)
 expression <- data.frame(GeneSymbol, FDR, p.value.adj, dm)
 gene.met <- starbursAnalysis(met, expression)
 starburstPlot(gene.met)
```

