---
title: "Glioma downstream  with TCGAbiolinks package"
author: " Antonio Colaprico, Tiago Chedraoui Silva, Luciano Garofano, 
Catharina Olsen, Davide Garolini, Claudia Cava, Isabella Castiglioni,
Houtan Noushmehr, Gianluca Bontempi, Michele Ceccarelli"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    number_sections: no
    toc_depth: 2
    highlight: haddock

references:
- id: ref1
  title: Orchestrating high-throughput genomic analysis with Bioconductor
  author: 
  - family: Huber, Wolfgang and Carey, Vincent J and Gentleman, Robert and Anders, Simon and Carlson, Marc and Carvalho, Benilton S and Bravo, Hector Corrada and Davis, Sean and Gatto, Laurent and Girke, Thomas and others
    given:
  journal: Nature methods
  volume: 12
  number: 2
  pages: 115-121
  issued:
    year: 2015

- id: ref2
  title: GC-content normalization for RNA-Seq data
  author: 
  - family: Risso, Davide and Schwartz, Katja and Sherlock, Gavin and Dudoit, Sandrine
    given:
  journal: BMC bioinformatics
  volume: 12
  number: 1
  pages: 480
  issued:
    year: 2011 
    
- id: ref3
  title: Evaluation of statistical methods for normalization and differential expression in mRNA-Seq experiments
  author: 
  - family: Bullard, James H and Purdom, Elizabeth and Hansen, Kasper D and Dudoit, Sandrine
    given:
  journal: BMC bioinformatics
  volume: 11
  number: 1
  pages: 94
  issued:
    year: 2010 


vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dpi = 150)
```

```{r, echo = FALSE,hide=TRUE, message=FALSE}
devtools::load_all(".")
```

# Introduction 

This vignette shows a complete workflow of the TCGAbiolinks package. 
The code is divided in 4 case study:

* 1. Expression pipeline (BRCA)
* 2. Expression pipeline (GBM)
* 3. Methylation pipeline - LGG/OV
* 4. Elmer pipeline - KIRC

# Parameters definition
```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
PlatformCancer <- "IlluminaHiSeq_RNASeqV2"
dataType <- "rsem.genes.results"
pathGBM<- "../dataGBM"
pathLGG <- "../dataLGG"

library(BiocInstaller)
useDevel()  # we need Devel for SummarizedExperiment package
library(SummarizedExperiment)
library(TCGAbiolinks)
```

# Case study n. 1: Pan Cancer downstream analysis BRCA

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
library(TCGAbiolinks)

cancer <- "BRCA"
PlatformCancer <- "IlluminaHiSeq_RNASeqV2"
dataType <- "rsem.genes.results"
pathCancer <- paste0("../data",cancer)

# Result....Function.....parameters...p1...pn..................................time execution

datQuery <- TCGAquery(tumor = cancer, platform = PlatformCancer, level = "3")  # time = 0.093s
lsSample <- TCGAquery_samplesfilter(query = datQuery)
dataSubt <- TCGAquery_subtype(tumor = cancer)
dataSmTP <- TCGAquery_SampleTypes(barcode = lsSample$IlluminaHiSeq_RNASeqV2,
                                  typesample = "TP")
dataSmTN <- TCGAquery_SampleTypes(barcode = lsSample$IlluminaHiSeq_RNASeqV2,
                                  typesample ="NT")
dataClin <- TCGAquery_clinic(cancer = cancer,
                             clinical_data_type = "clinical_patient")

            TCGAdownload(data = datQuery, path = pathCancer, type = dataType,
                         samples =c(dataSmTP,dataSmTN))

dataAssy <- TCGAprepare(query = datQuery, 
                        dir = pathCancer, 
                        type = dataType, 
                        save = TRUE, 
                        summarizedExperiment = TRUE, 
                        samples = c(dataSmTP,dataSmTN),
                        filename = paste0(cancer,"_",PlatformCancer,".rda"))

dataPrep <- TCGAanalyze_Preprocessing(object = dataAssy,cor.cut = 0.6)
dataNorm <- TCGAanalyze_Normalization(tabDF = dataPrep,
                                      geneInfo = geneInfo,
                                      method = "gcContent")
dataFilt <- TCGAanalyze_Filtering(tabDF = dataNorm, qnt.cut = 0.25)
dataDEGs <- TCGAanalyze_DEA(mat1 = dataFilt[,dataSmTN],
                            mat2 = dataFilt[,dataSmTP],
                            Cond1type = "Normal",
                            Cond2type = "Tumor",
                            fdr.cut = 0.01 ,
                            logFC.cut = 1,
                            method = "glmLRT")

dataSurv<-TCGAanalyze_SurvivalKM(clinical_patient = clinical_patient_Cancer,
                                 dataGE = dataFilt,
                                 
dataSurv<-TCGAanalyze_SurvivalKM(clinical_patient = dataClin,
                                dataGE = dataFilt,
                                Genelist = rownames(dataDEGs)[1:50],
                                Survresult = FALSE,
                                ThreshTop = 0.67,
                                ThreshDown = 0.33)
```


# Case study n. 2: Pan Cancer downstream analysis LGG

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
library(TCGAbiolinks)
library(genefilter)
cancer <- "LGG"
PlatformCancer <- "IlluminaHiSeq_RNASeqV2"
dataType <- "rsem.genes.results"
pathCancer <- paste0("../data",cancer)

# Result....Function.....parameters...p1...pn..................................time execution

datQuery <- TCGAquery(tumor = cancer, platform = PlatformCancer, level = "3")  # time = 0.093s
lsSample <- TCGAquery_samplesfilter(query = datQuery)
dataSubt <- TCGAquery_subtype(tumor = cancer)
dataSmTP <- TCGAquery_SampleTypes(barcode = lsSample$IlluminaHiSeq_RNASeqV2,
                                  typesample = "TP")
dataClin <- TCGAquery_clinic(cancer = cancer,
                             clinical_data_type = "clinical_patient")

            TCGAdownload(data = datQuery, path = pathCancer, type = dataType,
                         samples = dataSmTP )

dataAssy <- TCGAprepare(query = datQuery, 
                        dir = pathCancer, 
                        type = dataType, 
                        save = TRUE, 
                        summarizedExperiment = TRUE, 
                        samples = dataSmTP,
                        filename = paste0(cancer,"_",PlatformCancer,".rda"))

dataPrep <- TCGAanalyze_Preprocessing(object = dataAssy,cor.cut = 0.6)        # time = 13.028s
dataNorm <- TCGAanalyze_Normalization(tabDF = dataPrep,
                                      geneInfo = geneInfo,
                                      method = "gcContent")                  # time = 165.577s 
dataFilt <- varFilter(dataNorm, var.func = IQR,  
                      var.cutoff = 0.75,
                      filterByQuantile = TRUE)

normCounts <- dataFilt
geData <- t(log(1 + normCounts, 2))

dim(geData)

eta <- .05
foldChange <- 1
#  filter n.1
filter <- apply(geData, 2, function(x) sum(quantile(x, probs = c(1 - eta, eta)) * c(1, -1)))
fivenum(filter)
sum(filter > foldChange)
geData <- geData[, which(filter > foldChange)]

#  filter n.2
filter <- apply(geData, 2, function(x) prod(quantile(x, probs =  c(1 - eta, eta)) - 10) < 0)
sum(filter)

geData <- geData[, which(filter)]
dim(geData)
genesSelectedForClustering <- colnames(geData)

system.time(sHc <- hclust(ddist <- dist(geData), method = "ward.D2"))      # time = 1.270 )

library("ConsensusClusterPlus")

ans <- ConsensusClusterPlus(ddist, maxK = 7, pItem = 0.9, reps=1000
                            , title="mc_consensus_k7_1000"
                            , clusterAlg = "hc"
                            , innerLinkage = "ward.D2"
                            , finalLinkage = "complete"
                            , plot = 'pdf', writeTable = TRUE)
```



# Step 1: Query, download and prepare data LGG and GBM  

## Step 1.1: Query, download and prepare data LGG and GBM  rnaseqV2

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
library(TCGAbiolinks)
library(SummarizedExperiment)

# CancerPreprocess is a function that integrates query, download and prepare

CancerPreprocess <- function(cancer,PlatformCancer,pathCancer, nsamples = 0){
    query <- TCGAquery(tumor = cancer, platform = PlatformCancer, level = "3")
    listSamples <- TCGAquery_samplesfilter(query)
    listSamplesTP <- TCGAquery_SampleTypes(listSamples$IlluminaHiSeq_RNASeqV2,"TP") 
    
    if( nsamples!=0) {listSamplestoDown <- listSamplesTP[1:nsamples]}
    else{listSamplestoDown <- listSamplesTP}
   
    TCGAdownload(query, path = pathCancer, type = dataType,
             samples =listSamplestoDown )
  
    Cancer_assay <- TCGAprepare(query,dir = pathCancer,type = dataType, save = TRUE, 
                    filename = paste(cancer,"_",PlatformCancer,".rda",sep=""),
                    summarizedExperiment = FALSE)
      Cancer_matrix <- Cancer_assay[,listSamplesTP]
    # Cancer_matrix <- SummarizedExperiment::assay(Cancer_assay,"raw_counts")
return(Cancer_matrix)
}

```

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
library(TCGAbiolinks)
library(SummarizedExperiment)

# To test the functions use nsamples = 10 or 5
# GBM_matrix <- CancerPreprocess("gbm" ,PlatformCancer ,pathGBM,nsamples = 100)
# LGG_matrix <- CancerPreprocess("lgg" ,PlatformCancer ,pathLGG,nsamples = 20)

GBM_matrix <- CancerPreprocess("gbm" ,PlatformCancer ,pathGBM)
LGG_matrix <- CancerPreprocess("lgg" ,PlatformCancer ,pathLGG)
save(GBM_matrix, LGG_matrix, file =  "./dataLGG_GBM.Rdata")
```

## Step 1.2: Query, download and prepare data LGG and GBM  rnaseqV2

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
#-----------------------------------
# STEP 1: Search, download, prepare |
#-----------------------------------
# 1.1 - Methylation
# ----------------------------------
query.met <- TCGAquery(tumor = c("gbm","lgg"),
                       platform = c("HumanMethylation27",
                                    "HumanMethylation450"),
                       level = 3, version = list(c("HumanMethylation450","GBM",5)))

TCGAdownload(query.met, path = "/dados/ibm/comparing/biolinks/gbmlgg/")

met.gbm.lgg.27.450 <- TCGAprepare(query = query.met,
                                  dir = "/dados/ibm/comparing/biolinks/gbmlgg/",
                                  save = TRUE, filename = "gbmlgg.rda")
```


# Step 2: Analysis
## Step 2.1: Normalization and Differentially expression analysis

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
dataCancer <- cbind(GBM_matrix, LGG_matrix)

dataNorm <- TCGAanalyze_Normalization(dataCancer, geneInfo,method = "gcContent")
#dataFilt <- TCGAanalyze_Filtering(dataNorm, 0.25)
> dim(dataNorm)
[1] 20500   672

dataFilt <- dataNorm
dataDEGs <- TCGAanalyze_DEA(dataFilt[,colnames(GBM_matrix)], 
                            dataFilt[,colnames(LGG_matrix)],
                            "GBM", "LGG",method = "glmLRT")

save(dataDEGs, file ="./LGG_GBM_DEGs.Rdata")
```

## Step 2.2: mean Methylation by samples, DMR analysis, Starburst plot

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
#--------------------------------------------
# STEP 2: Analysis
#--------------------------------------------
# 2.1 - Mean methylation per patien
# -------------------------------------------
TCGAvisualize_meanMethylation(aux,
                              groupCol = "cluster.meth",subgroupCol = "disease",
                              group.legend  = "Groups", subgroup.legend = "Tumor",
                              print.pvalue = TRUE)

#------------------------------------------
# 2.2 - Volcano plot
# -----------------------------------------

# na.omit
aux <- subset(aux,subset = (rowSums(is.na(assay(aux))) == 0))

# Volcano plot
aux <- TCGAanalyze_DMR(aux, groupCol = "cluster.meth", p.cut = 10^-10, diffmean.cut = 0.25)
save(aux,file = "gbmlggwithpvalues.rda")

#------------------------------------------
# 2.3 - Starburst plot
# -----------------------------------------
startburst <- TCGAvisualize_starburst(aux, dataDEGs,"LGm1","LGm2", p.cut = 10^-10)
```


# Case study n. 4: Elmer pipeline - KIRC

One of the biggest problems related to the study data is the preparation phase,
which often consists of successive 
steps in order to prepare it to a format acceptable by certain algorithms and software.

With the object of assisting users in this arduous step, TCGAbiolinks offers in 
data preparation stage, the toPackage argument, which aims to prepare the data 
in order to obtain the correct format for different packages.

An example of another package to perform analysis is ELMER. We will present this
case study the study KIRC by TCGAbiolinks and ELMER integration.

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
#-----------------------------------
# STEP 1: Search, download, prepare |
#-----------------------------------
# Step 1.1 download methylation data|
# -----------------------------------
path <-  "."
query <- TCGAquery(tumor = "KIRC",level = 3, platform = "HumanMethylation450")
TCGAdownload(query, path = path)
kirc.met <- TCGAprepare(query,dir = path, save = TRUE,
                        filename = "metKirc.rda",
                        toPackage = "ELMER")

# Step 1.2 download expression data
query.rna <- TCGAquery(tumor="KIRC",level=3, platform="IlluminaHiSeq_RNASeqV2")
TCGAdownload(query.rna,path=path,type = "rsem.genes.normalized_results")

kirc.exp <- TCGAprepare(query.rna, dir=path, save = TRUE,
                        type = "rsem.genes.normalized_results",
                        filename = "expKirc.rda",toPackage = "ELMER")

#-----------------------------------
# STEP 2: ELMER integration         |
#-----------------------------------
# Step 2.1 prepare mee object       |
# -----------------------------------
library(ELMER)
library(parallel)

geneAnnot <- txs()
geneAnnot$GENEID <- paste0("ID",geneAnnot$GENEID)
geneInfo <- promoters(geneAnnot,upstream = 0, downstream = 0)
probe <- get.feature.probe()
mee <- fetch.mee(meth = kirc.met, exp = kirc.exp, TCGA = TRUE,
                 probeInfo = probe, geneInfo = geneInfo)

#--------------------------------------
# STEP 3: Analysis                     |
#--------------------------------------
# Step 3.1: Get diff methylated probes |
#--------------------------------------
Sig.probes <- get.diff.meth(mee ,cores=detectCores(),
                            dir.out ="kirc",diff.dir="hypo",
                            pvalue = 0.01)


#--------------------------------------------------
# Step 3.2: Identifying putative probe-gene pairs |
#--------------------------------------------------
# Collect nearby 20 genes for Sig.probes
nearGenes <- GetNearGenes(TRange=getProbeInfo(mee, probe=Sig.probes),
                          cores=detectCores(),
                          geneAnnot=getGeneInfo(mee))

# Identify significant probe-gene pairs
Hypo.pair <- get.pair(mee=mee,
                      probes=Sig.probes,
                      nearGenes=nearGenes,
                      permu.dir="./kirc/permu",
                      dir.out="./kirc/",
                      cores=detectCores(),
                      label= "hypo",
                      permu.size=10000,
                      Pe = 0.01)

Sig.probes.paired <- fetch.pair(pair=Hypo.pair,
                                probeInfo = getProbeInfo(mee),
                                geneInfo = getGeneInfo(mee))

#-------------------------------------------------------------
# Step 3.3: Motif enrichment analysis on the selected probes |
#-------------------------------------------------------------
enriched.motif <- get.enriched.motif(probes=Sig.probes.paired,
                                     dir.out="kirc", label="hypo",
                                     background.probes = probe$name)

#-------------------------------------------------------------
# Step 3.4: Identifying regulatory TFs                        |
#-------------------------------------------------------------
TF <- get.TFs(mee=mee,
              enriched.motif=enriched.motif,
              dir.out="kirc",
              cores=detectCores(), label= "hypo")

```

From this analysis it is possible to verify the relation between a probe 
and nearby genes. The result of this is show by the ELMER scatter plot.

```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center",hide=TRUE, message=FALSE,warning=FALSE}
library(png)
library(grid)
img <- readPNG("case4_elmer.png")
grid.raster(img)
```

******
### Session Information
******
```{r sessionInfo}
sessionInfo()
```

# References
