---
title: "Glioma downstream  with TCGAbiolinks package"
author: " Antonio Colaprico, Tiago Chedraoui Silva, Luciano Garofano, 
Catharina Olsen, Davide Garolini, Claudia Cava, Isabella Castiglioni,
Houtan Noushmehr, Gianluca Bontempi, Michele Ceccarelli"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    number_sections: no
    toc_depth: 2
    highlight: haddock

references:
- id: ref1
  title: Orchestrating high-throughput genomic analysis with Bioconductor
  author: 
  - family: Huber, Wolfgang and Carey, Vincent J and Gentleman, Robert and Anders, Simon and Carlson, Marc and Carvalho, Benilton S and Bravo, Hector Corrada and Davis, Sean and Gatto, Laurent and Girke, Thomas and others
    given:
  journal: Nature methods
  volume: 12
  number: 2
  pages: 115-121
  issued:
    year: 2015

- id: ref2
  title: GC-content normalization for RNA-Seq data
  author: 
  - family: Risso, Davide and Schwartz, Katja and Sherlock, Gavin and Dudoit, Sandrine
    given:
  journal: BMC bioinformatics
  volume: 12
  number: 1
  pages: 480
  issued:
    year: 2011 
    
- id: ref3
  title: Evaluation of statistical methods for normalization and differential expression in mRNA-Seq experiments
  author: 
  - family: Bullard, James H and Purdom, Elizabeth and Hansen, Kasper D and Dudoit, Sandrine
    given:
  journal: BMC bioinformatics
  volume: 11
  number: 1
  pages: 94
  issued:
    year: 2010 


vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(dpi = 150)
```

```{r, echo = FALSE,hide=TRUE, message=FALSE}
devtools::load_all(".")
```

# Introduction 

This vignette shows a complete workflow of the TCGAbiolinks package. 
The code is divided in 4 case study:

* 1. Expression pipeline (BRCA)
* 2. Expression pipeline (GBM)
* 3. Methylation pipeline - LGG/OV
* 4. Elmer pipeline - KIRC

# Parameters definition
```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
PlatformCancer <- "IlluminaHiSeq_RNASeqV2"
dataType <- "rsem.genes.results"
pathGBM<- "../dataGBM"
pathLGG <- "../dataLGG"

library(BiocInstaller)
useDevel()  # we need Devel for SummarizedExperiment package
library(SummarizedExperiment)
library(TCGAbiolinks)
```

# TCGAbiolinks Case study n. 1: Pan Cancer downstream analysis BRCA

## Step 1.1: Query, download and prepare data BRCA rnaseqV2

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
library(TCGAbiolinks)
library(SummarizedExperiment)

cancer <- "BRCA"
PlatformCancer <- "IlluminaHiSeq_RNASeqV2"
dataType <- "rsem.genes.results"
pathCancer <- paste0("../data",cancer)

datQuery <- TCGAquery(tumor = cancer, platform = PlatformCancer, level = "3")
lsSample <- TCGAquery_samplesfilter(datQuery)
dataSubt <- TCGAquery_subtype(tumor = cancer)
dataSmTP <- TCGAquery_SampleTypes(lsSample$IlluminaHiSeq_RNASeqV2,"TP")[1:50]
dataSmTN <- TCGAquery_SampleTypes(lsSample$IlluminaHiSeq_RNASeqV2,"NT")[1:50]
dataClin <- TCGAquery_clinic("brca","clinical_patient")

            TCGAdownload(query, path = pathCancer, type = dataType,
                         samples =c(dataSmTP,dataSmTN))

dataAssy <- TCGAprepare(query,dir = pathCancer,type = dataType, save = TRUE,
                    filename = paste(cancer,"_",PlatformCancer,".rda",sep=""),
                    summarizedExperiment = TRUE,samples = c(dataSmTP,dataSmTN))

dataPrep <- TCGAanalyze_Preprocessing(dataAssy)
dataNorm <- TCGAanalyze_Normalization(assay(dataAssy), geneInfo,method = "gcContent")
dataFilt <- TCGAanalyze_Filtering(TableRnaseq = dataNorm,QuantileThresh = 0.25)
dataDEGs <- TCGAanalyze_DEA(dataFilt[,listSamplesNT], 
                            dataFilt[,listSamplesTP],
                            "Normal", "Tumor",method = "glmLRT")
dataFDRs <- cbind(dataDEGs$table, FDR = p.adjust(dataDEGs$table$PValue, "fdr"))
dataFDRs <- dataFDRs[dataFDRs$FDR < 0.01,]
dataFDRs <- dataFDRs[abs(dataFDRs$logFC) > 1,]

library( "genefilter")
library("RColorBrewer")
library(gplots)

topVarGenes <- head( order( rowVars( dataFilt[rownames(dataFDRs),]), decreasing=TRUE ), 20 )

heatmap.2( as.matrix(dataFilt[ topVarGenes, ], scale="row",
        trace="none", dendrogram="column",
        col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(255)))
```



# Step 1: Query, download and prepare data LGG and GBM  

## Step 1.1: Query, download and prepare data LGG and GBM  rnaseqV2

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
library(TCGAbiolinks)
library(SummarizedExperiment)

# CancerPreprocess is a function that integrates query, download and prepare

CancerPreprocess <- function(cancer,PlatformCancer,pathCancer, nsamples = 0){
    query <- TCGAquery(tumor = cancer, platform = PlatformCancer, level = "3")
    listSamples <- TCGAquery_samplesfilter(query)
    listSamplesTP <- TCGAquery_SampleTypes(listSamples$IlluminaHiSeq_RNASeqV2,"TP") 
    
    if( nsamples!=0) {listSamplestoDown <- listSamplesTP[1:nsamples]}
    else{listSamplestoDown <- listSamplesTP}
   
    TCGAdownload(query, path = pathCancer, type = dataType,
             samples =listSamplestoDown )
  
    Cancer_assay <- TCGAprepare(query,dir = pathCancer,type = dataType, save = TRUE, 
                    filename = paste(cancer,"_",PlatformCancer,".rda",sep=""),
                    summarizedExperiment = FALSE)
      Cancer_matrix <- Cancer_assay[,listSamplesTP]
    # Cancer_matrix <- SummarizedExperiment::assay(Cancer_assay,"raw_counts")
return(Cancer_matrix)
}

```

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
library(TCGAbiolinks)
library(SummarizedExperiment)

# To test the functions use nsamples = 10 or 5
# GBM_matrix <- CancerPreprocess("gbm" ,PlatformCancer ,pathGBM,nsamples = 100)
# LGG_matrix <- CancerPreprocess("lgg" ,PlatformCancer ,pathLGG,nsamples = 20)

GBM_matrix <- CancerPreprocess("gbm" ,PlatformCancer ,pathGBM)
LGG_matrix <- CancerPreprocess("lgg" ,PlatformCancer ,pathLGG)
save(GBM_matrix, LGG_matrix, file =  "./dataLGG_GBM.Rdata")
```

## Step 1.2: Query, download and prepare data LGG and GBM  rnaseqV2

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
#-----------------------------------
# STEP 1: Search, download, prepare |
#-----------------------------------
# 1.1 - Methylation
# ----------------------------------
query.met <- TCGAquery(tumor = c("gbm","lgg"),
                       platform = c("HumanMethylation27",
                                    "HumanMethylation450"),
                       level = 3, version = list(c("HumanMethylation450","GBM",5)))

TCGAdownload(query.met, path = "/dados/ibm/comparing/biolinks/gbmlgg/")

met.gbm.lgg.27.450 <- TCGAprepare(query = query.met,
                                  dir = "/dados/ibm/comparing/biolinks/gbmlgg/",
                                  save = TRUE, filename = "gbmlgg.rda")
```


# Step 2: Analysis
## Step 2.1: Normalization and Differentially expression analysis

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
dataCancer <- cbind(GBM_matrix, LGG_matrix)

dataNorm <- TCGAanalyze_Normalization(dataCancer, geneInfo,method = "gcContent")
#dataFilt <- TCGAanalyze_Filtering(dataNorm, 0.25)
> dim(dataNorm)
[1] 20500   672

dataFilt <- dataNorm
dataDEGs <- TCGAanalyze_DEA(dataFilt[,colnames(GBM_matrix)], 
                            dataFilt[,colnames(LGG_matrix)],
                            "GBM", "LGG",method = "glmLRT")

save(dataDEGs, file ="./LGG_GBM_DEGs.Rdata")
```

## Step 2.2: mean Methylation by samples, DMR analysis, Starburst plot

```{r,eval=FALSE,echo=TRUE,message=FALSE,warning=FALSE}
#--------------------------------------------
# STEP 2: Analysis
#--------------------------------------------
# 2.1 - Mean methylation per patien
# -------------------------------------------
TCGAvisualize_meanMethylation(aux,
                              groupCol = "cluster.meth",subgroupCol = "disease",
                              group.legend  = "Groups", subgroup.legend = "Tumor",
                              print.pvalue = TRUE)

#------------------------------------------
# 2.2 - Volcano plot
# -----------------------------------------

# na.omit
aux <- subset(aux,subset = (rowSums(is.na(assay(aux))) == 0))

# Volcano plot
aux <- TCGAanalyze_DMR(aux, groupCol = "cluster.meth", p.cut = 10^-10, diffmean.cut = 0.25)
save(aux,file = "gbmlggwithpvalues.rda")

#------------------------------------------
# 2.3 - Starburst plot
# -----------------------------------------
startburst <- TCGAvisualize_starburst(aux, dataDEGs,"LGm1","LGm2", p.cut = 10^-10)
```


******
### Session Information
******
```{r sessionInfo}
sessionInfo()
```

# References
