---
title: "Working with TCGAbiolinks package"
author: " Antonio Colaprico, Tiago Chedraoui Silva, Luciano Garofano, 
Catharina Olsen, Davide Garolini, Claudia Cava, Isabella Castiglioni,
Houtan Noushmehr, Gianluca Bontempi, Michele Ceccarelli"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true

vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

```{r, echo = FALSE}
library(TCGAbiolinks)
```
## Introduction 

This document describes how to search, download and analyze TCGA data using the
TCGAbiolinks package.

## Searching data using TCGAQuery
 You can easily search TCGA samples using the `TCGAQuery` function.
 Using a summary of filters as used in the TCGA portal, the function works
 with the following 
 parameters:

* tumor 
* platform 
* added.since (format: mm/dd/YYYY)
* added.up.to (format: mm/dd/YYYY)
* listSample
* level
* center
* samples 

### Searching by tumor
You can filter the search by tumor using the tumor parameter.

```{r, eval = TRUE}
query <- TCGAQuery(tumor = "gbm")
```

If you don't remember the tumor name, or if you have incorrectly typed it. 
It will provide you with all the tumor names in TCGA.  Also the names can be seen in the
help pages `?TCGAQuery`

```{r, eval = TRUE}
query <- TCGAQuery(tumor = "")
```

### Searching by level
You can filter the search by level	"1", "2", "3" or "mage-tab"
```{r, eval = TRUE}
query <- TCGAQuery(tumor = "gbm", level = 3)
query <- TCGAQuery(tumor = "gbm", level = 2)
query <- TCGAQuery(tumor = "gbm", level = 1)
query <- TCGAQuery(tumor = "gbm", level = "mage-tab")
```

### Searching by date
You can filter the search by date using the added.since and added.up.to parameters.
For the moment, the format of date accepted is mm/dd/YYYY.

```{r, eval = TRUE}
# Get all gbm data produced in 2013
query <- TCGAQuery(tumor = "gbm", added.since = "01/01/2013", added.up.to = "12/31/2013")
```

### Searching by platform
You can filter the search by platform using the platform parameter.

```{r, eval = TRUE}
query <- TCGAQuery(tumor = "gbm", platform = "IlluminaHiSeq_RNASeqV2")
```

If you don't remember the platform, or if you have incorrectly typed it. 
It will provide you with all the platforms names in TCGA. Also the names can be seen in the
help pages `?TCGAQuery`

```{r, eval = TRUE}
query <- TCGAQuery(tumor = "gbm", platform = "")
```

### Searching by center
You can filter the search by center using the center parameter.

```{r, eval = TRUE}
query <- TCGAQuery(tumor = "gbm", center = "mskcc.org")
```

If you don't remember the center or if you have incorrectly typed it. 
It will provide you with all the center names in TCGA. 

```{r, eval = TRUE}
query <- TCGAQuery(tumor = "gbm", center = "")
```

### Searching by samples
You can filter the search by samples using the samples parameter.
You can give a list of barcodes or only one barcode. These barcode can be partial barcodes.

```{r, eval = TRUE}
# You can define a list of samples to query and download providing relative TCGA barcodes.
listSamples <- c("TCGA-E9-A1NG-11A-52R-A14M-07","TCGA-BH-A1FC-11A-32R-A13Q-07",
                 "TCGA-A7-A13G-11A-51R-A13Q-07","TCGA-BH-A0DK-11A-13R-A089-07",
                 "TCGA-E9-A1RH-11A-34R-A169-07","TCGA-BH-A0AU-01A-11R-A12P-07",
                 "TCGA-C8-A1HJ-01A-11R-A13Q-07","TCGA-A7-A13D-01A-13R-A12P-07",
                 "TCGA-A2-A0CV-01A-31R-A115-07","TCGA-AQ-A0Y5-01A-11R-A14M-07")

# Query all available platforms with a list of barcode 
query <- TCGAQuery(samples = listSamples)

# Query with a partial barcode 
query <- TCGAQuery(samples = "TCGA-61-1743-01A")
```

### Examples

Some search examples are shown below:
```{r, eval = TRUE}
query <- TCGAQuery(tumor = "gbm", added.since = "01/01/2013", added.up.to = "12/31/2013")

query <- TCGAQuery(tumor = c("gbm","lgg"), platform = c("HumanMethylation450","HumanMethylation27"))

query <- TCGAQuery(tumor = "gbm", platform = "HumanMethylation450", level = "3")

query <- TCGAQuery(samples = "TCGA-61-1743-01A-01D")

query <- TCGAQuery(samples = "TCGA-61-1743-01A-01D-0649-04", level = 3)

query <- TCGAQuery(samples = "TCGA-61-1743-01A-01D-0649-04", tumor = "OV", platform = "CGH-1x1M_G4447A")
```


Some search examples are shown below
```{r, eval = TRUE}
query <- TCGAQuery(tumor = "brca",level = 3)
querySamples <- samplesfilter(query)

query <- TCGAQuery(tumor = "brca", level = 3, platform = "IlluminaHiSeq_RNASeqV2")

matSamples <- TCGAintegrate(querySamples)

colnames(matSamples) <- paste("p",1:length(names(querySamples)), 
                              names(querySamples),sep=".")
rownames(matSamples) <- paste("p",1:length(names(querySamples)),
                              names(querySamples),sep=".")
colnames(matSamples) <- substr( colnames(matSamples), 1,4)


```

The result from TCGAintegrate is shown below:

```{r, eval = TRUE, echo = FALSE}
colnames(matSamples) <- paste("p",1:length(names(querySamples)), 
                              names(querySamples),sep=".")
rownames(matSamples) <- paste("p",1:length(names(querySamples)),
                              names(querySamples),sep=".")
colnames(matSamples) <- substr( colnames(matSamples), 1,4)

knitr::kable(matSamples, digits = 2, 
             caption = "Table common samples among platforms from TCGAQuery",
             row.names = TRUE)


```


Query version for a specific platform for example IlluminaHiSeq_RNASeqV2
```{r, eval = FALSE}
library(TCGAbiolinks)

BRCA_RNASeqV2_version <- TCGAVersion(tumor = "brca", 
                                     platform = "illuminahiseq_rnaseqv2")

```

The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
BRCA_RNASeqV2_version <- BRCA_RNASeqV2_version[order(BRCA_RNASeqV2_version$SizeMbyte,decreasing=TRUE),]
knitr::kable(BRCA_RNASeqV2_version, digits = 2, 
             caption = "Table with version, number of samples and size (Mbyte) of BRCA IlluminaHiSeq_RNASeqV2 Level 3", 
             row.names = FALSE)
```



## Downloading data with TCGADownload
You can easily download data using the `TCGADownload` function.

The arguments are:
* data	The TCGAQuery output
* path	location to save the files. Default: "."
* type	Filter the files to download by type
* samples	List of samples to download
* quiet	Supress output messages? Default: FALSE

Example of use:
```{r, eval = FALSE}
# get all samples from the query and save them in the TCGA folder
# samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results
# samples to normalize later
TCGADownload(query, path = "data", type = "rsem.genes.results")

```

Comment: The function will structure the folders to save the data as:
_Path given by the user/Experiment folder_

## Preparing the data with TCGAPrepare
You can easily read your downloaded data using the `TCGAPrepare` function.
This function will prepare data matrices for downstream analysis.

The arguments are:
* query	Data frame as the one returned from TCGAQuery
* dir	Directory with the files
* type	File to prepare.
* save	Save a rda object with the prepared object? Default: FALSE
* filename Name of the rda object that will be saved if `save` is TRUE
* toPackage Name of the package to prepare the data specific to that package. 

Example of use:
```{r, eval = FALSE}
# get all samples from the query and save them in the TCGA folder
# samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results
# samples to normalize later
TCGAPrepare(query, dir = "data", save = TRUE, filename = "myfile.rda")
```

### Preparing the data with TCGAPrepare - toPackage 
This section will show how to integrate TCGAbiolinks with other packages.
Our intention is to provide as many integrations as possible.

The example below shows how to use TCGAbiolinks with ELMER package 
(expression/methylation analysis). The TCGAPrepare for the methylation data 
will Removing probes with NA values in more than 0.80% samples and remove the 
anottation data, fot the expression data it will take the log2(expression + 1)
of the expression matrix in order to To linearize the relation between 
methylation and expressionm also it will prepare the rownames as the 
specified by the package.

```{r, eval = FALSE}
########################### Get tumor samples with TCGAbiolinks
library(TCGAbiolinks)
query <- TCGAQuery(tumor = "GBM",level = 3, platform = "HumanMethylation450")
TCGADownload(query,path = "TCGA/450k")
met <- TCGAPrepare(query,dir = "TCGA/450k",
                   save = TRUE,
                   filename = "met.rda",
                   toPackage = "ELMER")

query.rna <- TCGAQuery(tumor="GBM",level=3, platform="IlluminaHiSeq_RNASeqV2")
TCGADownload(query.rna,path="TCGA/rna",type = "rsem.genes.normalized_results")
exp <- TCGAPrepare(query.rna, dir="TCGA/rna", save = TRUE, filename = "exp.rda",toPackage = "ELMER")

######################### To EMLER
library(ELMER)

################### genne annotation
geneAnnot <- txs()
geneAnnot$GENEID <- paste0("ID",geneAnnot$GENEID)
geneInfo <- promoters(geneAnnot,upstream = 0, downstream = 0)
######################## probe
probe <- get.feature.probe()

mee.gbm.glial.with.exp <- fetch.mee(meth = gbm.glial.m,
                                exp = exp,
                                probeInfo = probe,
                                TCGA = TRUE,
                                geneInfo = geneInfo)
```



## Clinical data
You can retrive clinical data using the `clinic` function. The parameters of this
function is: 

* cancer ("OV","BRCA","GBM", etc)
* clinical_data_type ("clinical_patient","clinical_drug", etc)

A full list of cancer and clinical data type can be found in the help of the function.

```{r, eval = FALSE}
# Get clinical data
clinical_brca_data <- clinic("brca","clinical_patient")
clinical_uvm_data_bio <- clinic("uvm","biospecimen_normal_control")
clinical_brca_data_bio <- clinic("brca","biospecimen_normal_control")
clinical_brca_data <- clinic("brca","clinical_patient")
```

Also, some functions to work with clinical data are provided.
For example the function `gender_BRCA` returns the list of barcodes of female patients.

```{r, eval = FALSE}
bar <- c("TCGA-G9-6378-02A-11R-1789-07", "TCGA-CH-5767-04A-11R-1789-07",  
         "TCGA-G9-6332-60A-11R-1789-07", "TCGA-G9-6336-01A-11R-1789-07",
         "TCGA-G9-6336-11A-11R-1789-07", "TCGA-G9-7336-11A-11R-1789-07",
         "TCGA-G9-7336-04A-11R-1789-07", "TCGA-G9-7336-14A-11R-1789-07",
         "TCGA-G9-7036-04A-11R-1789-07", "TCGA-G9-7036-02A-11R-1789-07",
         "TCGA-G9-7036-11A-11R-1789-07", "TCGA-G9-7036-03A-11R-1789-07",
         "TCGA-G9-7036-10A-11R-1789-07", "TCGA-BH-A1ES-10A-11R-1789-07",
         "TCGA-BH-A1F0-10A-11R-1789-07", "TCGA-BH-A0BZ-02A-11R-1789-07",
         "TCGA-B6-A0WY-04A-11R-1789-07", "TCGA-BH-A1FG-04A-11R-1789-08",
         "TCGA-D8-A1JS-04A-11R-2089-08", "TCGA-AN-A0FN-11A-11R-8789-08",
         "TCGA-AR-A2LQ-12A-11R-8799-08", "TCGA-AR-A2LH-03A-11R-1789-07",
         "TCGA-BH-A1F8-04A-11R-5789-07", "TCGA-AR-A24T-04A-55R-1789-07",
         "TCGA-AO-A0J5-05A-11R-1789-07", "TCGA-BH-A0B4-11A-12R-1789-07",
         "TCGA-B6-A1KN-60A-13R-1789-07", "TCGA-AO-A0J5-01A-11R-1789-07",
         "TCGA-AO-A0J5-01A-11R-1789-07", "TCGA-G9-6336-11A-11R-1789-07",
         "TCGA-G9-6380-11A-11R-1789-07", "TCGA-G9-6380-01A-11R-1789-07")

S <- SampleTypes(bar,"TP")

# Retrieve multiple tissue types  NOT FROM THE SAME PATIENTS
SS <- MultiSampleTypes(bar,c("TP","NB"))

# Retrieve multiple tissue types  FROM THE SAME PATIENTS
SSS <- MatchedCoupledSampleTypes(bar,c("NT","TP"))

# Get clinical data
clinical_brca_data <- clinic("brca","clinical_patient")

stage_BRCA(bar,"stage_IIIA",clinical_brca_data)

gender_BRCA(bar,"FEMALE",clinical_brca_data)

ER_status_BRCA(bar,"Positive",clinical_brca_data)

ER_status_BRCA(bar,"Negative",clinical_brca_data)

PR_status_BRCA(bar,"Positive",clinical_brca_data)

PR_status_BRCA(bar,"Negative",clinical_brca_data)

HER_status_BRCA(bar,"Positive",clinical_brca_data)

HER_status_BRCA(bar,"Negative",clinical_brca_data)
```

## [Gene Expression] TCGA query, download, prepare list of a barcode
You can easily search TCGA samples, download and prepare a matrix of gene expression.
```{r, eval = FALSE}

# You can define a list of samples to query and download providing relative TCGA barcodes.

listSamples <- c("TCGA-E9-A1NG-11A-52R-A14M-07","TCGA-BH-A1FC-11A-32R-A13Q-07",
                 "TCGA-A7-A13G-11A-51R-A13Q-07","TCGA-BH-A0DK-11A-13R-A089-07",
                 "TCGA-E9-A1RH-11A-34R-A169-07","TCGA-BH-A0AU-01A-11R-A12P-07",
                 "TCGA-C8-A1HJ-01A-11R-A13Q-07","TCGA-A7-A13D-01A-13R-A12P-07",
                 "TCGA-A2-A0CV-01A-31R-A115-07","TCGA-AQ-A0Y5-01A-11R-A14M-07")

# Query all available platforms with a list of barcode 
query <- TCGAQuery(samples = listSamples)

# Query platform IlluminaHiSeq_RNASeqV2 with a list of barcode 
query <- TCGAQuery(tumor = "brca", samples = listSamples, 
  platform = "IlluminaHiSeq_RNASeqV2", level = "3")

# Download a list of barcodes with platform IlluminaHiSeq_RNASeqV2
TCGADownload(query, path = "dataBrca", type = "rsem.genes.results",
             samples = listSamples)

# Prepare expression matrix with gene id in rows and samples (barcode) in columns
# rsem.genes.results as values
BRCAMatrix <- TCGAPrepare(query,"dataBrca")
```

The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
dataGE <- dataBRCA[sample(rownames(dataBRCA),10),sample(colnames(dataBRCA),7)]

knitr::kable(dataGE, digits = 2, 
             caption = "Example of a matrix of gene expression (10 genes in rows and 7 samples in columns)",
             row.names = TRUE)
```



## [CNV] TCGA query, download, prepare list of a barcode
You can easily search TCGA samples, download and prepare a matrix of gene expression.
```{r, eval = FALSE}

# Define a list of samples to query and download providing relative TCGA barcodes.
samplesList <- c("TCGA-02-0046-10A-01D-0182-01", 
                 "TCGA-02-0052-01A-01D-0182-01", 
                 "TCGA-02-0033-10A-01D-0182-01", 
                 "TCGA-02-0034-01A-01D-0182-01", 
                 "TCGA-02-0007-01A-01D-0182-01")

# Query platform Genome_Wide_SNP_6 with a list of barcode 
query <- TCGAQuery(tumor = "gbm", level = 3, platform = "Genome_Wide_SNP_6")

# Download a list of barcodes with platform Genome_Wide_SNP_6
TCGADownload(query, path = "samples")

# Prepare matrix 
GBM_CNV <- TCGAPrepare(query, dir = "samples", type = ".hg19.seg.txt")
```


## TCGA Downstream Analysis TCGAprepare, TCGAanalyze, TCGAvisualize
You can prepare gene expression from TCGA data using the `TCGAprepare`
function.
 
```{r, eval = FALSE}
# Downstream analysis using gene expression data  
# TCGA samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results

library(TCGAbiolinks)

# dataBRCA in TCGAbiolinks package is a table from TCGA BRCA [10 samples] and comes from 
# BRCAMatrix <- TCGAPrepare(query,"dataBrca") from above example
# dataBRCA <- BRCAMatrix

# normalization of genes
dataNorm <- TCGAbiolinks::RnaSeqNormalization(dataBRCA, geneInfo)

# quantile filter of genes
dataFilt <- RnaSeqFilt(dataNorm, 0.25)

# selection of normal samples "NT"
samplesNT <- MultiSampleTypes(colnames(dataFilt), typesample = c("NT"))

# selection of tumor samples "TP"
samplesTP <- MultiSampleTypes(colnames(dataFilt), typesample = c("TP"))


```

## TCGA analyze and visualize (Diff.expr.analysis (DEA))
You can analyze gene expression from TCGA data using the `TCGAanalyze` 
function.
 
```{r, eval = FALSE}
# Downstream analysis using gene expression data  
# TCGA samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results
# save(dataBRCA, geneInfo , file = "dataGeneExpression.rda")
library(TCGAbiolinks)

# Diff.expr.analysis (DEA)
dataDEGs <- DEArnaSEQ(dataFilt[,samplesNT], dataFilt[,samplesTP],
                      "Normal", "Tumor")

# DEGs filter by abs(logFC) >=1
dataDEGsFilt <- dataDEGs[abs(dataDEGs$logFC) >= 1,]

# DEGs table with expression values in normal and tumor samples
dataDEGsFiltLevel <- CreateTabLevel(dataDEGsFilt,"Tumor","Normal",
                                    dataFilt[,samplesTP],dataFilt[,samplesNT])

```
The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
dataDEGsFiltLevel$FDR <- format(dataDEGsFiltLevel$FDR, scientific = TRUE)
knitr::kable(dataDEGsFiltLevel[1:10,], digits = 2,
             caption = "Table DEGs after DEA",row.names = FALSE)
```



## TCGA analyze and visualize (Enrichment Analysis)
You can analyze gene expression from TCGA data using the `TCGAanalyze` function.
 
```{r, eval = FALSE}
library(TCGAbiolinks)
# Enrichment Analysis EA
# Gene Ontology (GO) and Pathway enrichment by DEGs list
Genelist <- rownames(dataDEGsFiltLevel)

system.time(ansEA <- EAcomplete(TFname="DEA genes Normal Vs Tumor",Genelist))

# Enrichment Analysis EA (TCGAVisualize)
# Gene Ontology (GO) and Pathway enrichment barPlot

EAbarplot(tf = rownames(ansEA$ResBP), 
           GOBPTab = ansEA$ResBP,
           GOCCTab = ansEA$ResCC,
           GOMFTab = ansEA$ResMF,
           PathTab = ansEA$ResPat,
           nRGTab = Genelist, 
           nBar = 10)

```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("EAplot.png")
grid.raster(img)
```

## TCGA analyze (Enrichment Analysis)
You can analyze gene expression from TCGA data using the `TCGAanalyze` function.
 
```{r, eval = FALSE}
library(TCGAbiolinks)

# normalization of genes
dataNorm <- TCGAbiolinks::RnaSeqNormalization(dataBRCA, geneInfo)

# quantile filter of genes
dataFilt <- RnaSeqFilt(dataNorm, 0.25)

# Principal Component Analysis plot for ntop selected DEGs
plotPCAforGroups(dataFilt,dataDEGsFiltLevel, ntopgenes = 200)

# boxplot of normalized data
#sampleGenes <- rownames(dataDEGsFilt[dataDEGsFilt$logFC >=1,])[1:20]
#boxplot(log(dataBRCA[sampleGenes,]), las = 2)
#boxplot(log(dataFilt[sampleGenes,]), las = 2)


```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("PCAtop200DEGs.png")
grid.raster(img)
```

## TCGA analyze and visualize (Survival Analysis)
You can analyze gene expression and clinical data from TCGA data using the `TCGAanalyze` function.
 
```{r, eval = FALSE}
library(TCGAbiolinks)
# Survival Analysis SA

clinical_patient_Cancer <- clinic("brca","clinical_patient")
dataBRCAcomplete <- log2(BRCA_rnaseqv2)

tokenStop<- 1

tabSurvKMcomplete <- NULL

for( i in 1: round(nrow(dataBRCAcomplete)/100)){
message( paste( i, "of ", round(nrow(dataBRCAcomplete)/100)))
tokenStart <- tokenStop
tokenStop <-100*i
tabSurvKM<-SurvivalKMunivariate(clinical_patient_Cancer,dataBRCAcomplete,Genelist = rownames(dataBRCAcomplete)[tokenStart:tokenStop],
                                        Survresult = F,ThreshTop=0.67,ThreshDown=0.33)

tabSurvKMcomplete <- rbind(tabSurvKMcomplete,tabSurvKM)
}

tabSurvKMcomplete <- tabSurvKMcomplete[tabSurvKMcomplete$pvalue < 0.01,]
tabSurvKMcomplete <- tabSurvKMcomplete[!duplicated(tabSurvKMcomplete$mRNA),]
rownames(tabSurvKMcomplete) <-tabSurvKMcomplete$mRNA
tabSurvKMcomplete <- tabSurvKMcomplete[,-1]
tabSurvKMcomplete <- tabSurvKMcomplete[order(tabSurvKMcomplete$pvalue, decreasing=F),]

tabSurvKMcompleteDEGs <- tabSurvKMcomplete[rownames(tabSurvKMcomplete) %in% dataDEGsFiltLevel$mRNA,]



```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}
tabSurvKMcompleteDEGs$pvalue <- format(tabSurvKMcompleteDEGs$pvalue, scientific = TRUE)
knitr::kable(tabSurvKMcompleteDEGs[1:10,], digits = 2,
             caption = "Table KM-survival genes after SA",row.names = TRUE)
```

## TCGA analyze and visualize (Survival Analysis Cox Regression and dnet package)
You can analyze gene expression and clinical data from TCGA data using the `TCGAanalyze` function.
 
```{r, eval = FALSE}
library(TCGAbiolinks)
# Survival Analysis SA

clinical_patient_Cancer <- clinic("brca","clinical_patient")
dataBRCAcomplete <- log2(BRCA_rnaseqv2)

tokenStop<- 1

tabSurvKMcomplete <- NULL

for( i in 1: round(nrow(dataBRCAcomplete)/100)){
message( paste( i, "of ", round(nrow(dataBRCAcomplete)/100)))
tokenStart <- tokenStop
tokenStop <-100*i
tabSurvKM<-SurvivalKMunivariate(clinical_patient_Cancer,dataBRCAcomplete,Genelist = rownames(dataBRCAcomplete)[tokenStart:tokenStop],Survresult = F,ThreshTop=0.67,ThreshDown=0.33)

tabSurvKMcomplete <- rbind(tabSurvKMcomplete,tabSurvKM)
}

tabSurvKMcomplete <- tabSurvKMcomplete[tabSurvKMcomplete$pvalue < 0.01,]
tabSurvKMcomplete <- tabSurvKMcomplete[!duplicated(tabSurvKMcomplete$mRNA),]
rownames(tabSurvKMcomplete) <-tabSurvKMcomplete$mRNA
tabSurvKMcomplete <- tabSurvKMcomplete[,-1]
tabSurvKMcomplete <- tabSurvKMcomplete[order(tabSurvKMcomplete$pvalue, decreasing=F),]

tabSurvKMcompleteDEGs <- tabSurvKMcomplete[rownames(tabSurvKMcomplete) %in% dataDEGsFiltLevel$mRNA,]

tflist <- EAGenes[EAGenes$Family == "transcription regulator","Gene"]
tabSurvKMcomplete_onlyTF <- tabSurvKMcomplete[rownames(tabSurvKMcomplete) %in% tflist,]

TabCoxNet <- SurvivalCoxNET(clinical_patient,dataBRCAcomplete,Genelist = rownames(tabSurvKMcomplete_onlyTF),scoreConfidence = 700,titlePlot = "SurvivalCoxNET Example")



```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}

library(png)
library(grid)
img <- readPNG("SurvivalCoxNETOutput.png")
grid.raster(img)
```


## TCGA investigate
Find most studied TFs in pubmed related to a specific cancer, disease, or tissue
 
```{r, eval = FALSE}
# First perform DEGs with TCGAanalyze
# See previous section 
library(TCGAbiolinks)

# Select only transcription factors (TFs) from DEGs
TFs <- EAGenes[EAGenes$Family =="transcription regulator",]
TFs_inDEGs <- intersect(TFs$Gene, dataDEGsFiltLevel$mRNA )
dataDEGsFiltLevelTFs <- dataDEGsFiltLevel[TFs_inDEGs,]

# Order table DEGs TFs according to Delta decrease
dataDEGsFiltLevelTFs <- dataDEGsFiltLevelTFs[order(dataDEGsFiltLevelTFs$Delta,decreasing = TRUE),]

# Find Pubmed of TF studied related to cancer
tabDEGsTFPubmed <- TCGAinvestigate("breast", dataDEGsFiltLevelTFs, topgenes = 10)
                 
```

The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
knitr::kable(tabDEGsTFPubmed, digits = 2, caption = "Table with most studied TF in pubmed related to a specific cancer",row.names = FALSE)
```

## TCGA social
The `TCGASocial` function has two type of searches, one that searches for
most downloaded packages in CRAN or BioConductor and one that searches the most related question in biostar 



### TCGA social with BioConductor
Find most downloaded packages in CRAN or BioConductor
 
```{r, eval = TRUE}
library(TCGAbiolinks)

# Define a list of package to find number of downloads
listPackage <-c("limma","edgeR","survcomp")

tabPackage <- TCGAsocial(siteToFind ="bioconductor.org",listPackage)


# define a keyword to find in support.bioconductor.org returing a table with suggested packages
tabPackageKey <- TCGAsocial(siteToFind ="support.bioconductor.org" ,KeyInfo = "tcga")
```

The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
listPackage <-c("limma","edgeR","survcomp")
tabPackage <- TCGAsocial(siteToFind="bioconductor.org", listPackage)

knitr::kable(head(tabPackage), digits = 2, caption = "Table with number of downloads about a list of packages",row.names = FALSE)

knitr::kable(tabPackage2, digits = 2, caption = "Find most related question in support.bioconductor.org with keyword = tcga",row.names = FALSE)

```

### TCGA social with Biostar
Find most related question in biostar.
 
```{r, eval = FALSE}
library(TCGAbiolinks)

# Find most related question in biostar with TCGA
tabPackage1 <- TCGAsocial(siteToFind ="biostars.org",KeyInfo = "TCGA")

# Find most related question in biostar with package
tabPackage2 <- TCGAsocial(siteToFind ="biostars.org",KeyInfo = "package")

```

The result is shown below:

```{r, eval = TRUE, echo = FALSE}
library(TCGAbiolinks)
# Find most related question in biostar with TCGA
#tabPackage <- TCGAsocial(siteToFind ="biostars.org",KeyInfo = "package")

knitr::kable(head(tabPackage1), digits = 2, caption = "Find most related question in biostar with TCGA",row.names = FALSE)

knitr::kable(tabPackage2, digits = 2, caption = "Find most related question in biostar with package",row.names = FALSE)
```

## TCGA downstream Analysis
 
Some examples for the integration of TCGA data are shown below
```{r, eval = FALSE}

library(TCGAbiolinks)
query <- TCGAQuery(tumor = "brca",level = 3)

querySamples <- samplesfilter(query)

query <- TCGAQuery(tumor = "brca",level = 3, platform = "IlluminaHiSeq_RNASeqV2")

```

Comment: As these terms need to be exact, we validate user input before searching.

## TCGA Downstream Analysis Integration 
You can easily prepare gene expression from TCGA data using the `TCGAPrepare` function.
 
```{r, eval = FALSE}
library(TCGAbiolinks)
library(genefilter)
library(clue)

BRCArnaseqV2 <- dataBRCA
BRCArnaseqV2MostVar <- varFilter(BRCArnaseqV2, var.func = IQR,  var.cutoff = 0.75, filterByQuantile = TRUE)

wData <- t(BRCArnaseqV2MostVar)
ddist <- dist(wData, method = "euclidean")
sHc <- hclust(ddist, method = "ward.D")

plot(sHc, labels = FALSE, main ="BRCA Cancer cluster dendrogram all samples",
   xlab = "Samples with relative group color",sub="")

rect.hclust(sHc, k=3, border="red")
tabCluster <- as.matrix(cutree(sHc, k = 3))
colnames(tabCluster)<-"Cluster"
tabCluster<-cbind(Sample = rownames(tabCluster),Color = rownames(tabCluster), tabCluster)
tabCluster<-as.data.frame(tabCluster)
tabCluster<-tabCluster[order(tabCluster$Cluster,decreasing = FALSE),]
tabCluster<-as.data.frame(tabCluster)
tabCluster$Color<-as.character(tabCluster$Color)

ccol <- palette()[1 + 1:3]

for( cc in 1:3){
  tabCluster[tabCluster[, "Cluster"] == cc, "Color"] <- ccol[cc]
}

tabCluster <- tabCluster[sHc$labels, ]

rug(which(tabCluster[sHc$order, "Color"] == "blue"), col = "blue", lwd = 3)
rug(which(tabCluster[sHc$order, "Color"] == "green3"), col = "green3", lwd = 3)
rug(which(tabCluster[sHc$order, "Color"] == "red"), col = "red", lwd = 3)
```

The result is shown below:
```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("BRCA_cluster.png")
grid.raster(img)
```


```{r, eval = FALSE}
library(TCGAbiolinks)

##### Differential analysis
GroupBlueData <- BRCArnaseqV2[, as.character(tabCluster[tabCluster$Color == "blue", "Sample"])]
GroupGreen3Data <- BRCArnaseqV2[, as.character(tabCluster[tabCluster$Color == "green3", "Sample"])]
GroupRedData <- BRCArnaseqV2[, as.character(tabCluster[tabCluster$Color == "red", "Sample"])]

DEGsBlue <- DEArnaSEQ(cbind(GroupGreen3Data, GroupRedData),
                      GroupBlueData, "GroupOther", "GroupBlue")
DEGsGreen3 <- DEArnaSEQ(cbind(GroupBlueData, GroupRedData),
                        GroupGreen3Data, "GroupOther", "GroupGreen3")
DEGsRed <- DEArnaSEQ(cbind(GroupBlueData, GroupGreen3Data), 
                     GroupRedData, "GroupOther", "GroupRed")


dataDEGs <- DEArnaSEQ(dataFilt[,samplesNT], dataFilt[,samplesTP], "Normal", "Tumor")

# DEGs filter by abs(logFC) >=1
dataDEGsFilt <- dataDEGs[abs(dataDEGs$logFC) >= 1,]

dataDEGsFiltLevel <- CreateTabLevel(dataDEGsFilt,"Tumor","Normal",
                                    [,samplesTP],dataFilt[,samplesNT])

DEGsBlueLevel <- CreateTabLevel(DEGsBlue, "GroupBlue", "GroupOther", 
  GroupBlueData, cbind(GroupGreen3Data, GroupRedData), typeOrder = TRUE)
DEGsGreen3Level <- CreateTabLevel(DEGsGreen3, "GroupGreen3", "GroupOther",
  GroupGreen3Data, cbind(GroupBlueData, GroupRedData), typeOrder = TRUE)
DEGsRedLevel <- CreateTabLevel(DEGsRed, "GroupRed", "GroupOther",
  GroupRedData, cbind(GroupBlueData, GroupGreen3Data), typeOrder = TRUE)

blueDEGs <- DEGsBlueLevel[DEGsBlueLevel$FDR < 0.01 & DEGsBlueLevel$logFC >= 1, ]
blueDEGs <- blueDEGs[order(blueDEGs$FDR), ]
green3DEGs <- DEGsGreen3Level[DEGsGreen3Level$FDR < 0.01 & DEGsGreen3Level$logFC >= 1, ]
green3DEGs <- green3DEGs[order(green3DEGs$FDR), ]
redDEGs <- DEGsRedLevel[DEGsRedLevel$FDR < 0.01 & DEGsRedLevel$logFC >= 1, ]
redDEGs <- redDEGs[order(redDEGs$FDR), ]

blueDEGsSpec <- blueDEGs[setdiff(rownames(blueDEGs), 
                                 union(rownames(green3DEGs), rownames(redDEGs))), ]
green3DEGsSpec <- green3DEGs[setdiff(rownames(green3DEGs), 
                                     union(rownames(blueDEGs), rownames(redDEGs))), ]
redDEGsSpec <- redDEGs[setdiff(rownames(redDEGs), 
                               union(rownames(blueDEGs), rownames(green3DEGs))), ]

blueDEGsSpec <- blueDEGsSpec[1:50, ]
green3DEGsSpec <- green3DEGsSpec[1:50, ]
redDEGsSpec <- redDEGsSpec[1:50, ]

tabCluster <- tabCluster[order(tabCluster$Color), ]

MfiltQuantileOrdered <- BRCArnaseqV2[c(rownames(blueDEGsSpec), rownames(green3DEGsSpec), 
  rownames(redDEGsSpec)), rownames(tabCluster)]

MRactivity <- t(MfiltQuantileOrdered)

HMactivity <- MRactivity
thresholdquantile<-0.75
HMactivity[ HMactivity >= quantile(HMactivity,thresholdquantile)]<- quantile(HMactivity,thresholdquantile)

summary(as.vector(HMactivity))
quantile(HMactivity, 0.15); quantile(HMactivity, 0.85)
HMactivity[HMactivity <= quantile(HMactivity, 0.15)] <- quantile(HMactivity, 0.15)
HMactivity[HMactivity >= quantile(HMactivity, 0.85)] <- quantile(HMactivity, 0.85)

column_annotation <- matrix(" ", nrow = nrow(HMactivity), ncol = 1)
column_annotation[, 1] <- tabCluster$Color

row_annotation <- matrix(" ", nrow = 1, ncol = ncol(HMactivity))
row_annotation[1, ] <- c(rep("blue", nrow(blueDEGsSpec)),
                         rep("green3", nrow(green3DEGsSpec)), 
                         rep("red", nrow(redDEGsSpec)))
library("GMD")

png("BRCA_heatmap.png", width = 1200, height = 800)
heatmap.3(t(HMactivity), ColSideColors = column_annotation, RowSideColors = row_annotation,
          key = FALSE, Colv = NA, Rowv = NA,
          scale = "none", 
          #col = greenred(75), 
          dendrogram = "none",
          #labRow = NA, labCol = NA,
          margins = c(1, 6), side.height.fraction = 0.25, keysize = 1.4, cexRow = 1.6)
dev.off()
```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("BRCA_heatmap.png")
grid.raster(img)
```


#### TCGA downstream methylation analysis

Some downstream analysis from methylation data can be done with TCGAbiolinks.
An example is shown below. We search, download and prepare data from the 
HumanMethylation450 platform.

```{r, eval = FALSE}
library(TCGAbiolinks)
samples <- c("TCGA-06-0125-01A-01D-A45W-05","TCGA-06-0152-01A-02D-A45W-05",
             "TCGA-06-0171-01A-02D-A45W-05","TCGA-06-0190-01A-01D-A45W-05",
             "TCGA-06-0210-01A-01D-A45W-05","TCGA-06-0211-01A-01D-A45W-05",
             "TCGA-06-0221-01A-01D-A45W-05","TCGA-14-0736-01A-01D-A45W-05",
             "TCGA-19-0957-01C-01D-A45W-05","TCGA-19-1389-01A-01D-A45W-05",
             "TCGA-14-1402-01A-01D-A45W-05","TCGA-19-4065-01A-01D-2004-05",
             "TCGA-06-0125-02A-11D-2004-05","TCGA-06-0152-02A-01D-2004-05",
             "TCGA-06-0171-02A-11D-2004-05","TCGA-06-0190-02A-01D-2004-05",
             "TCGA-06-0210-02A-01D-2004-05","TCGA-06-0211-02A-02D-2004-05",
             "TCGA-06-0221-02A-11D-2004-05","TCGA-14-0736-02A-01D-2004-05",
             "TCGA-19-0957-02A-11D-2004-05","TCGA-19-1389-02A-21D-2004-05",
             "TCGA-14-1402-02A-01D-2004-05","TCGA-19-4065-02A-11D-2004-05")

query <- TCGAQuery(tumor = "gbm", platform = "HumanMethylation450",
                   level = "3", samples = samples)
TCGADownload(query,path = "exampleData")

#-------------------------- Preparing data and metadata
# met: data frame with probe info (probe id, ge) and beta values
# beta: data frame with only beta values (probe vs patient)
# probe: data frame with only probe info(probe vs patient)
# Conclusion: met = probe +d beta
# met.md = patient metadata
met <- TCGAPrepare(query, dir="exampleData")
met <-  met[ met$Chromosome != "X" &  met$Chromosome != "Y", ]
met <- na.omit(met)
probe  <- met[,1:4]
beta   <- met[,5:ncol(met)]


query <- TCGAQuery(tumor = "gbm", platform = "bio", level = "2")
TCGADownload(query, path = "exampleData")
met.md <- TCGAPrepare(query, dir="exampleData")

samples <- colnames(beta)
idx <- is.element(met.md$bcr_patient_barcode,strtrim(samples,12))
met.md <- met.md[idx,]
```

We divided the data into groups 
in order to analyze the data. Using the metadata, it is possible to create a
survival plot with the function survivalPlot as follows:

```{r, eval = FALSE}
# random split of pacients into groups
met.md$cluster <- c(rep("group1",nrow(met.md)/4),
                    rep("group2",nrow(met.md)/4),
                    rep("group3",nrow(met.md)/4),
                    rep("group4",nrow(met.md)-3*(floor(nrow(met.md)/4))))
rownames(met.md) <- met.md$bcr_patient_barcode

#---------------------- survival
survivalPlot(met.md)
```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("survival.png")
grid.raster(img)
```

Using the data and calculating the mean methylation per group, it is possible to create a
mean methylation boxplot with the function met.mean.boxplot as follows:

```{r, eval = FALSE}
#----------------------mean methylation
met.mean <- data.frame(apply(beta, 2, mean, na.rm = TRUE))
colnames(met.mean) <- "avg"
met.mean$patient <- strtrim(rownames(met.mean), 12)
aux <- merge(met.mean, met.md, by.x = "patient", by.y = "bcr_patient_barcode")
met.mean.boxplot(aux)
```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("meanmet.png")
grid.raster(img)
```

The final example is a volcano plot 
that compares the methylation data between the two groups.

```{r, eval = FALSE}
#----------------------- calculate.pvalues (just an example)
beta.t <-  data.frame(t(beta))
# TODO verify class of the object
pvalues <- calculate.pvalues(beta.t,c(1,3,5,7,9,11),c(2,4,6,8,10,12))

#------------- volcano plot
probe$p.value <- pvalues[,1]
probe$p.value.adj <- pvalues[,2]
diffmean <- diffmean(beta[,c(1,3,5,7,9,11)],beta[,c(2,4,6,8,10,12)])
probe <- cbind(probe,diffmean)
hypo.hyper <- volcanoPlot(data, p.cut = 0.85)
```

The output will be an graph such as the figure below. The function will 
return the most relevant values.

```{r, fig.width=6, fig.height=4, echo = FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("volcano.png")
grid.raster(img)
```

In the next part, we will analyze expression data and methylation data with 
starburst analysis.
```{r, eval = FALSE}
 FDR <- runif(nrow(met), 0, 1)
 p.value.adj <- runif(nrow(met), 0, 1)
 GeneSymbol <- met$Gene_Symbol
 dm <- runif(nrow(met), -3, 3)
 expression <- data.frame(GeneSymbol, FDR, p.value.adj, dm)
 gene.met <- starbursAnalysis(met, expression)
 starburstPlot(gene.met)
```

## Session Information
```{r, eval = TRUE}
sessionInfo()
```
