---
title: "Working with TCGAbiolinks package"
author: " Antonio Colaprico, Catharina Olsen, Luciano Garofano, Tiago Chedraoui Silva, Davide Garolini, Claudia Cava, Isabella Castiglioni, Houtan Noushmehr, Gianluca Bontempi, Michele Ceccarelli"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

This document describes how to search, download and analyze TCGA data using the TCGAbiolinks package.

## TCGA query
 You can easily search TCGA samples using the `tcgaQuery` function.
 Using a summary of filters as used in the TCGA portal, the function works with the following 
 parameters:

* tumor
* platform
* added.since
* added.up.to
* listSample
* level
* center

Some search examples are shown below
```{r, eval = FALSE}

library(TCGAbiolinks)
query <- TCGAQuery(tumor = "brca",level = 3)

querySamples <- samplesfilter(query)

query <- TCGAQuery(tumor = "brca",level = 3, platform = "IlluminaHiSeq_RNASeqV2")

```

Obs: As these terms needs to be exact, we validate the user input before searching.

## TCGA download
You can easily download roadmap data using the `tcga.download` function.
 
```{r, eval = FALSE}
# get all samples from the query and save into tcga folder
# samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results
# samples to normalize later
TCGADownload(query, path = "data", type = "rsem.genes.results")

```

Obs: The function will structure the folders to save the data as:
_Path given by the user/Experiment folder_

## TCGA download clinic data
 
```{r, eval = FALSE}

# Function clinic
# clinic(cancer,clinical_data_type)
# input: cancer, clinical_data_type 
# cancer = a character vector indicating cancer type to download
# Options include:
# ACC,BLCA,BRCA,CESC,COAD,DLBC,ESCA,GBM
# HNSC,KICH,KIRC,KIRP,LAML,LGG,LIHC,LUAD,
# LUSC,OV,PAAD,PRAD,READ,SARC,SKCM,STAD,THCA
# UCEC,UCS

# For information about cancer types: https://tcga-data.nci.nih.gov/tcga/
#
# clinical_data_type = a character vector indicating the types of 
# clinical data 
# 
# Options include:
# biospecimen_aliquot, biospecimen_analyte, biospecimen_cqcf, biospecimen_diagnostic_slides,
# biospecimen_normal_control, biospecimen_portion, biospecimen_protocol,
# biospecimen_sample, biospecimen_shipment_portion, biospecimen_slide,
# biospecimen_tumor_sample, clinical_cqcf, clinical_drug, clinical_follow_up_v1.5,
# clinical_follow_up_v2.1, clinical_follow_up_v4.0, clinical_follow_up_v4.0_nte,
# clinical_nte, clinical_omf_v4.0, clinical_patient, clinical_radiation 
# 
                       
clinic("gbm","clinical_patient")

```


## TCGA Downstream Analysis TCGAprepare, TCGAanalyze, TCGAvisualize
You can prepare gene expression from TCGA data using the `tcga.prepare` function.
 
```{r, eval = FALSE}
# Downstream analysis using gene expression data  
# TCGA samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results

#save(DAVID_BP_matrix, DAVID_CC_matrix, DAVID_MF_matrix, IPAGenes, listIPA_pathways, file = "dataEnrichmentAnalysis.rda")
# save(dataBRCA, geneInfo , file = "dataGeneExpression.rda")
library(TCGAbiolinks)

# normalization of genes
dataNorm <- TCGAbiolinks::RnaSeqNormalization(dataBRCA, geneInfo)

# quantile filter of genes
dataFilt <- RnaSeqFilt(dataNorm, 0.25)

# selection of normal samples "NT"
samplesNT <- MultiSampleTypes(colnames(dataFilt), typesample = c("NT"))

# selection of tumor samples "TP"
samplesTP <- MultiSampleTypes(colnames(dataFilt), typesample = c("TP"))


```

## TCGA analyze (Diff.expr.analysis (DEA))
You can analyze gene expression from TCGA data using the `tcga.analyze` function.
 
```{r, eval = FALSE}
# Downstream analysis using gene expression data  
# TCGA samples from IlluminaHiSeq_RNASeqV2 with type rsem.genes.results
# save(dataBRCA, geneInfo , file = "dataGeneExpression.rda")
library(TCGAbiolinks)

# Diff.expr.analysis (DEA)
dataDEGs <- DEArnaSEQ(dataFilt[,samplesNT], dataFilt[,samplesTP], "Normal", "Tumor")

# DEGs filter by abs(logFC) >=1
dataDEGsFilt <- dataDEGs[abs(dataDEGs$logFC) >= 1,]

# DEGs table with expression values in normal and tumor samples
dataDEGsFiltLevel <- CreateTabLevel(dataDEGsFilt,"Tumor","Normal",dataFilt[,samplesTP],dataFilt[,samplesNT])

knitr::kable(dataDEGsFiltLevel[1:10,], digits = 2, caption = "A table produced by printr.")
```
The result is shown below:

```{r, eval = TRUE}
library(TCGAbiolinks)
knitr::kable(dataDEGsFiltLevel[1:10,], digits = 2, caption = "A table produced by printr.")
```



## TCGA analyze (Enrichment Analysis)
You can analyze gene expression from TCGA data using the `tcga.analyze` function.
 
```{r, eval = FALSE}

# Enrichment Analysis EA
# Gene Ontology (GO) and Pathways enrichment by DEGs list
Genelist <- rownames(dataDEGsFiltLevel)

TimeUse(ansEA <- EAcomplete(TFname="DEA genes Normal Vs Tumor",Genelist))

# Enrichment Analysis EA (TCGAVisualize)
# Gene Ontology (GO) and Pathways enrichment IPA barPlot

IPAbarplot(tf = rownames(ansEA$ResBP), 
           GOBPTab = ansEA$ResBP,
           GOCCTab = ansEA$ResCC,
           GOMFTab = ansEA$ResMF,
           PathTab = ansEA$ResPat,
           nRGTab = Genelist, 
           nBar = 10)

# Principal Component Analysis Plot for ntop DEGs selected
plotPCAforGroups(dataFilt,dataDEGsFiltLevel, ntopgenes = 100)

# boxplot of data normalized 
#sampleGenes <- rownames(dataDEGsFilt[dataDEGsFilt$logFC >=1,])[1:20]
#boxplot(log(dataBRCA[sampleGenes,]), las = 2)
#boxplot(log(dataFilt[sampleGenes,]), las = 2)



```
The result is shown below:
```{r, fig.width=6, fig.height=4, echo=FALSE, fig.align="center"}
library(png)
library(grid)
img <- readPNG("EAplot.png")
grid.raster(img)
```

## TCGA investigate
Find most studied TF in pubmed related to a specific cancer, disease, or tissue
 
```{r, eval = FALSE}
# First perform DEGs with TCGAanalyze
# See upper vignette
library(TCGAbiolinks)

# Select only transcription factors (tfs) from DEGs
TFs <- IPAGenes[IPAGenes$Family =="transcription regulator",]
TFs_inDEGs <- intersect(TFs$Gene, dataDEGsFiltLevel$mRNA )
dataDEGsFiltLevelTFs <- dataDEGsFiltLevel[TFs_inDEGs,]

# Ordering table DEGs TFs according Delta decreasing
dataDEGsFiltLevelTFs <- dataDEGsFiltLevelTFs[order(dataDEGsFiltLevelTFs$Delta,decreasing=T),]

# Find Pubmed of TF studied related to cancer
tabDEGsTFPubmed <- TCGAinvestigate("breast", dataDEGsFiltLevelTFs, topgenes = 10)
                 
```

The result is shown below:

```{r, eval = TRUE}
library(TCGAbiolinks)
knitr::kable(tabDEGsTFPubmed, digits = 2, caption = "Table with most studied TF in pubmed related to a specific cancer",row.names = F)
```
